// @oracle-cx-commerce/actions/purchase-list/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createPurchaseList = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createPurchaseList'],
  description:
    'This API creates a purchase list for a given site, account and user context. Each purchase list should have a unique name for a given user. Dynamic Properties of purchase list can be added in the the request using the key value pair structure ("dynamicPropertyKey":"dynamicPropertyValue"). These property values can take the types of boolean, date, float, string, timestamp, enumerated, big string.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/create-purchase-list/input.json'
};

export const listPurchaseLists = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listPurchaseLists'],
  description:
    'This API is to get the collection of purchase lists. It can also be used to return the list on a searched string and in a sorted order against various properties of purchase list.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/list-purchase-lists/input.json'
};

export const getPurchaseList = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getPurchaseList'],
  description: 'This API is used to get the purchase list information for a given purchase list ID',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/get-purchase-list/input.json'
};

export const updatePurchaseList = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updatePurchaseList'],
  description:
    'This API update properties of a purchase list. Dynamic Properties of purchase list can be added in the the request using the key value pair structure ("dynamicPropertyKey":"dynamicPropertyValue"). These property values can take the types of boolean, date, float, string, timestamp, enumerated, big string.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/update-purchase-list/input.json'
};

export const deletePurchaseList = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deletePurchaseList'],
  description:
    'This API deletes the Purchase list information. Along with the puchase list, it’s sharing is also deleted.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/delete-purchase-list/input.json'
};

export const updatePurchaseListItems = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updatePurchaseListItems'],
  description:
    'This API is used to update the purchase list items. Basic operations include add, delete, update. add operation is considered as default if nothing is specified in the input payload.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/update-purchase-list-items/input.json'
};

export const checkPurchaseListEditAccess = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['checkPurchaseListEditAccess'],
  description: 'Checks the edit access of the current user for the purchase list.',
  input: '@oracle-cx-commerce/actions/purchase-list/schema/check-purchase-list-edit-access/input.json'
};
