// @oracle-cx-commerce/actions/payment/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addPayments = () => import('@oracle-cx-commerce/actions/payment/actions');
export const applyPayments = () => import('@oracle-cx-commerce/actions/payment/actions');
export const calculateRemainingBalance = () => import('@oracle-cx-commerce/actions/payment/actions');
export const clearInquireBalance = () => import('@oracle-cx-commerce/actions/payment/actions');
export const deleteAppliedPayment = () => import('@oracle-cx-commerce/actions/payment/actions');
export const getBillingCountries = () => import('@oracle-cx-commerce/actions/payment/actions');
export const getCardTypes = () => import('@oracle-cx-commerce/actions/payment/actions');
export const getPaymentGroup = () => import('@oracle-cx-commerce/actions/payment/actions');
export const getPaymentInitiatedOrder = () => import('@oracle-cx-commerce/actions/payment/actions');
export const getPaymentConfigurations = () => import('@oracle-cx-commerce/actions/payment/actions');
export const initiatePayment = () => import('@oracle-cx-commerce/actions/payment/actions');
export const inquireBalance = () => import('@oracle-cx-commerce/actions/payment/actions');
export const updateAppliedPayment = () => import('@oracle-cx-commerce/actions/payment/actions');
