// @oracle-cx-commerce/actions/payment/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
export const addPayments = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addPayments'],
  description: 'Add Payments. Used to authorize payments for order.',
  input: '@oracle-cx-commerce/actions/payment/schema/add-payments/input.json'
};

export const applyPayments = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['applyPayments'],
  description: 'Apply Payments. Used to apply payments on an order.',
  input: '@oracle-cx-commerce/actions/payment/schema/apply-payments/input.json'
};

export const calculateRemainingBalance = {
  packageId: '@oracle-cx-commerce/actions',
  description: 'Get remaining order amount for mixed currency orders.',
  input: '@oracle-cx-commerce/actions/payment/schema/calculate-remaining-balance/input.json'
};

export const clearInquireBalance = {
  packageId: '@oracle-cx-commerce/actions',
  description:
    'Clears inquired balances stored under the state for payment of types giftCards,loyaltyPoints or storeCredits.If no payload is passed all the stored balances are removed.',
  input: '@oracle-cx-commerce/actions/payment/schema/clear-inquire-balance/input.json',
  output: '@oracle-cx-commerce/actions/payment/schema/clear-inquire-balance/output.json'
};

export const deleteAppliedPayment = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteAppliedPayment'],
  description: 'Used to delete an applied payments on an order.',
  input: '@oracle-cx-commerce/actions/payment/schema/delete-applied-payment/input.json'
};

export const getBillingCountries = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getBillingCountries'],
  description: 'Get Billing Countries. Lists all the configured billing countries and regions.',
  input: '@oracle-cx-commerce/actions/payment/schema/get-billing-countries/input.json'
};

export const getCardTypes = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCardTypes']
};

export const getPaymentGroup = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getPaymentGroup'],
  description:
    'Get Payment Group. Get PaymentGroup authorization status an order. Optionally takes the x-ccasset-language header to get translated content in another language.',
  input: '@oracle-cx-commerce/actions/payment/schema/get-payment-group/input.json'
};

export const getPaymentInitiatedOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getPaymentInitiatedOrder'],
  description:
    'Returns payment initiated order that has been initiated with the payment type. Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.',
  input: '@oracle-cx-commerce/actions/payment/schema/get-payment-initiated-order/input.json'
};

export const getPaymentConfigurations = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getPaymentConfigurations']
};

export const initiatePayment = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['initiatePayment'],
  description: 'Initiate Payment. Used to intiate payment for web checkout, cash payment types.',
  input: '@oracle-cx-commerce/actions/payment/schema/initiate-payment/input.json'
};

export const inquireBalance = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['inquireBalance'],
  description: 'Inquire Balance. Used to inquired balance for giftCard, loyaltyPoints or storeCredit.',
  input: '@oracle-cx-commerce/actions/payment/schema/inquire-balance/input.json'
};

export const updateAppliedPayment = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateAppliedPayment'],
  description: 'Used to update applied payment on an order.',
  input: '@oracle-cx-commerce/actions/payment/schema/update-applied-payment/input.json'
};
