// @oracle-cx-commerce/actions/organization-address/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const getOrganizationDefaultAddresses = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrganizationDefaultAddresses'],
  description:
    'This operation is used to get the default shipping and billing address of a Organization/Account Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/get-organization-default-addresses/input.json'
};

export const getOrganizationAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrganizationAddress'],
  description: 'This endpoint is used to get address by its id of a Organization/Account in Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/get-organization-address/input.json'
};

export const updateOrganizationAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateOrganizationAddress'],
  description:
    'This endpoint is used to update addresses by its id of a Organization/Account in Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/update-organization-address/input.json'
};

export const addOrganizationAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addOrganizationAddress'],
  description: 'This endpoint is used to add addresses of a Organization/Account in Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/add-organization-address/input.json'
};

export const deleteOrganizationAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteOrganizationAddress'],
  description: 'This endpoint is used to delete address by its id of a Organization/Account in Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/delete-organization-address/input.json'
};

export const listOrganizationAddresses = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listOrganizationAddresses'],
  description: 'This operation is used to get the list of addresses of the current organization.',
  input: '@oracle-cx-commerce/actions/organization-address/schema/list-organization-addresses/input.json'
};
