// @oracle-cx-commerce/react-widgets/profile/profile-wish-list-social-sharing/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';

const configResourceKeys = [
  'configDisplayFacebookLabel',
  'configFacebookURLLabel',
  'configFacebookURLHelpText',
  'configDisplayPinterestLabel',
  'configPinterestURLLabel',
  'configPinterestURLHelpText',
  'configDisplayTwitterLabel',
  'configTwitterURLLabel',
  'configTwitterURLHelpText',
  'configDisplayEmailLabel'
];

const config = {
  properties: [
    {
      id: 'displayFacebook',
      type: 'booleanType',
      name: 'displayFacebook',
      labelResourceId: 'configDisplayFacebookLabel',
      defaultValue: true
    },
    {
      id: 'facebookURL',
      type: 'stringType',
      labelResourceId: 'configFacebookURLLabel',
      helpTextResourceId: 'configFacebookURLHelpText',
      defaultValue:
        'https://www.facebook.com/dialog/share?__facebookAppId__&display=popup&href=__wishlistURL__&redirect_uri=https://www.facebook.com'
    },
    {
      id: 'displayTwitter',
      type: 'booleanType',
      name: 'displayTwitter',
      labelResourceId: 'configDisplayTwitterLabel',
      defaultValue: true
    },
    {
      id: 'twitterURL',
      type: 'stringType',
      labelResourceId: 'configTwitterURLLabel',
      helpTextResourceId: 'configTwitterURLHelpText',
      defaultValue: 'https://twitter.com/share?url=__wishlistURL__&text=__textTwitterShare__'
    },
    {
      id: 'displayPinterest',
      type: 'booleanType',
      name: 'displayPinterest',
      labelResourceId: 'configDisplayPinterestLabel',
      defaultValue: true
    },
    {
      id: 'pinterestURL',
      type: 'stringType',
      labelResourceId: 'configPinterestURLLabel',
      helpTextResourceId: 'configPinterestURLHelpText',
      defaultValue: 'https://pinterest.com/pin/create/button/?url=__wishlistURL__&description=__textPinterestShare__'
    },
    {
      id: 'displayEmail',
      type: 'booleanType',
      name: 'displayEmail',
      labelResourceId: 'configDisplayEmailLabel',
      defaultValue: true
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
};

export default config;
