// @oracle-cx-commerce/endpoints/cart-shipping-group/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addShippingGroupsToCart = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/add-shipping-groups-to-cart');
export const deleteShippingGroupFromCart = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/delete-shipping-group-from-cart');
export const deleteShippingGroupsFromCart = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/delete-shipping-groups-from-cart');
export const getCartShippingGroup = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/get-cart-shipping-group');
export const getCartShippingGroups = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/get-cart-shipping-groups');
export const relateShippingGroupsToCartItems = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/relate-shipping-groups-to-cart-items');
export const updateCartShippingGroup = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/update-cart-shipping-group');
export const updateCartShippingGroups = () =>
  import('@oracle-cx-commerce/endpoints/cart-shipping-group/update-cart-shipping-groups');
