// @oracle-cx-commerce/endpoints/cart-shipping-group/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export {default as addShippingGroupsToCart} from '@oracle-cx-commerce/endpoints/cart-shipping-group/add-shipping-groups-to-cart/meta';
export {default as deleteShippingGroupFromCart} from '@oracle-cx-commerce/endpoints/cart-shipping-group/delete-shipping-group-from-cart/meta';
export {default as deleteShippingGroupsFromCart} from '@oracle-cx-commerce/endpoints/cart-shipping-group/delete-shipping-groups-from-cart/meta';
export {default as getCartShippingGroup} from '@oracle-cx-commerce/endpoints/cart-shipping-group/get-cart-shipping-group/meta';
export {default as getCartShippingGroups} from '@oracle-cx-commerce/endpoints/cart-shipping-group/get-cart-shipping-groups/meta';
export {default as relateShippingGroupsToCartItems} from '@oracle-cx-commerce/endpoints/cart-shipping-group/relate-shipping-groups-to-cart-items/meta';
export {default as updateCartShippingGroup} from '@oracle-cx-commerce/endpoints/cart-shipping-group/update-cart-shipping-group/meta';
export {default as updateCartShippingGroups} from '@oracle-cx-commerce/endpoints/cart-shipping-group/update-cart-shipping-groups/meta';
