// @oracle-cx-commerce/osf-config/app-config/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

module.exports = {
  cacheEnabled: true, // whether to enable cache or not
  cacheOptions: {
    validityPeriod: 2 * 60 * 1000, // validity period of the req to be in cache. If expired, the request goes to OCC for re-validation.
    ttl: 30 * 60, // if expired, the response against respective request stored in cache gets removed.
    max: 99999, // maximum objects to be stored in the cache.
    cacheableUrlPatterns: {
      // url patterns for which the urls should be cached. If the url doesn't fall under any of these, it is not cache-able.
      default: {
        urlPatterns: [
          '/ccstore/v1/collections',
          '/ccstore/v1/countries',
          '/ccstore/v1/files/',
          '/ccstore/v1/itemTypes',
          '/ccstore/v1/locations',
          '/ccstore/v1/merchant/cancelReasons',
          '/ccstore/v1/merchant/cloudConfiguration',
          '/ccstore/v1/merchant/paymentConfigurations',
          '/ccstore/v1/merchant/production-Experiments',
          '/ccstore/v1/merchant/production-Facebook',
          '/ccstore/v1/merchant/returnReasons',
          '/ccstore/v1/merchant/samlSettings',
          '/ccstore/v1/merchant/shopperSettings',
          '/ccstore/v1/payment/types',
          '/ccstore/v1/priceListGroups',
          '/ccstore/v1/prices/',
          '/ccstore/v1/products',
          '/ccstore/v1/productTypes',
          '/ccstore/v1/registry',
          '/ccstore/v1/shippingRegions',
          '/ccstore/v1/shippingMethods/',
          '/ccstore/v1/sites',
          '/ccstore/v1/xregistry'
        ]
      }
    }
  }
};
