// @oracle-cx-commerce/tools-util/auth-endpoints.js
/*
 ** Copyright (c) 2021 Oracle and/or its affiliates.
 */

// Schema: {
//   "endpointId": (parameters) => ({path: '...', method: '...', headers: {...}, body: [...]}),
//   ...
// }
const endpoints = {
  login: appKey => ({
    path: '/ccadmin/v1/login',
    method: 'post',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
      Accept: 'application/json',
      'Accept-Charset': 'utf-8',
      Authorization: `Bearer ${appKey}`
    },
    body: 'grant_type=client_credentials'
  }),
  logout: accessToken => ({
    path: '/ccadmin/v1/logout',
    method: 'post',
    headers: {
      'Content-Type': 'application/json; charset=utf-8',
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`
    },
    body: '{}'
  }),
  refresh: accessToken => ({
    path: '/ccadmin/v1/refresh',
    method: 'post',
    headers: {
      'Content-Type': 'application/json; charset=utf-8',
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`
    },
    body: '{}'
  })
};

module.exports = {
  ...endpoints
};
