// @oracle-cx-commerce/cli/templates/endpoints/string-utils.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
const {standardizePluginName} = require('@oracle-cx-commerce/cli/templates/utils');
const {dashCase, camelCase} = require('@oracle-cx-commerce/tools-util/strings');

function swaggerEndpointsIndexExportString(directoryNamePath) {
  return `export * from './${directoryNamePath}';\n`;
}
function swaggerEndpointsMetaExportString(directoryNamePath) {
  return `export * from './${directoryNamePath}/meta';\n`;
}
function endpointsIndexExportString(directoryName) {
  return `export const ${standardizePluginName(directoryName)} = () => import('./${directoryName}');\n`;
}
function endpointsMetaExportString(directoryName) {
  return `export {default as ${standardizePluginName(directoryName)}} from './${directoryName}/meta';\n`;
}
function endpointPluginIndexExport(operationId) {
  return `export const ${camelCase(operationId)} = () => import('./${dashCase(operationId)}');
`;
}
function endpointPluginMetaExport(operationId) {
  return `export {default as ${camelCase(operationId)}} from './${dashCase(operationId)}/meta';
`;
}
module.exports = {
  swaggerEndpointsIndexExportString,
  swaggerEndpointsMetaExportString,
  endpointsIndexExportString,
  endpointsMetaExportString,
  endpointPluginIndexExport,
  endpointPluginMetaExport
};
