// @oracle-cx-commerce/cli/templates/endpoints/transform-catalog-es5.js
/*
 ** Copyright (c) 2021 Oracle and/or its affiliates.
 */

/**
 * Converts a string into camel-case.
 *
 * @param      {string}  text    The text to convert.
 * @return     {string}  The text converted to camel-case format.
 */
function camelCase(text) {
  const capitalize = s => s[0].toUpperCase() + s.slice(1);
  if (!text || typeof text !== 'string' || !text.trim()) {
    return;
  }
  const matches = text.trim().match(/^[a-z]+|[a-z]+|[A-Z][a-z]+|\d+|[A-Z]+(?![a-z])/g);
  const capitalizedString = matches.map(word => capitalize(word)).join('');

  return capitalizedString[0].toLowerCase() + capitalizedString.slice(1);
}

/**
 * Filter all required parameters from endpoint's parameter array
 *
 * @param      {Array}  parameters  Array of endpoint's parameter
 * @return     {Array}  Array of required parameters
 */
function getRequiredParameters(parameters) {
  const requiredParameters = [];
  for (const entry of parameters) {
    if (entry.required && entry.in === 'path') {
      requiredParameters.push(entry);
    }
  }

  return requiredParameters;
}

function transformCatalogToEndpointMapping(rawCatalog) {
  const catalog = {};
  const {basePath = '', paths = {}} = rawCatalog;

  for (const [pathName, pathValue] of Object.entries(paths)) {
    for (const [methodName, methodValue] of Object.entries(pathValue)) {
      const {consumes, produces, parameters = [], description = 'No description available'} = methodValue;

      let {operationId = `${methodName}${basePath}${pathName}`} = methodValue;
      operationId = methodValue.operationId ? operationId : camelCase(operationId);

      const requiredParameters = parameters.length > 0 ? getRequiredParameters(parameters) : [];

      catalog[operationId] = {
        id: operationId,
        operationId,
        url: basePath + pathName.replace(/{.*?}/g, '{}'),
        method: methodName,
        description: description.replace(/[\n']/g, ''),
        requiredParameters
      };

      if (consumes) {
        catalog[operationId].requestType = consumes.join(',');
      }

      if (produces) {
        catalog[operationId].responseType = produces.join(',');
      }
    }
  }

  return catalog;
}
/**
 * Transform a swagger ApiCatalog into an endpoint mapping
 * @param {*} currentCatalog
 * @param {*} rawCatalog
 */
// eslint-disable-next-line no-undef
module.exports = (currentCatalog, rawCatalog) => {
  const transformedCatalog = transformCatalogToEndpointMapping(rawCatalog);

  return Object.assign(currentCatalog, transformedCatalog);
};
