// @oracle-cx-commerce/react-widgets/common/cookie-control/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

//List of cookies that will be preserved/not deleted when GDPR cookie consent is not granted
export const necessaryCookies = 'access_token,xdVisitorId,xvVisitId,OCStateData';

const configResourceKeys = ['labelTextCookiePolicyURL', 'webContentLabel'];

const config = mergeDefaultConfig({
  properties: [
    {
      id: 'cookiePolicyURL',
      type: 'stringType',
      labelResourceId: 'labelTextCookiePolicyURL',
      defaultValue: ''
    },
    {
      id: 'webContent',
      type: 'webContentType',
      labelResourceId: 'webContentLabel',
      defaultValue: '<p>Some Web Content Text</p>'
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
});

export default config;
