// @oracle-cx-commerce/react-widgets/cart/extended-cart-item-details/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import extendedCartItemConfig from '@oracle-cx-commerce/react-widgets/cart/extended-cart-item-details/config';
import {productDeliveryOptionsWidgetResourceKeys} from '@oracle-cx-commerce/react-widgets/product/product-delivery-options/meta';
import {cartItemDetailsWidgetResourceKeys} from '@oracle-cx-commerce/react-widgets/cart/cart-item-details/meta';
import {widgetResourceKeys as configuratorResourceKeys} from '@oracle-cx-commerce/react-widgets/product/product-configure/meta';

export const extendedCartItemDetailsWidgetResourceKeys = ['headingStorePickUpAt'];

export const widgetResourceKeys = [
  'actionHideDetails',
  'actionShowDetails',
  'textPrice',
  ...productDeliveryOptionsWidgetResourceKeys,
  ...cartItemDetailsWidgetResourceKeys,
  ...extendedCartItemDetailsWidgetResourceKeys,
  ...configuratorResourceKeys
];

export const ExtendedCartItemDetails = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: extendedCartItemConfig,
  availableToAllPages: false,
  pageTypes: ['cart'],
  requiresContext: ['cart_context']
};
