// @oracle-cx-commerce/actions/order-approval/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const approveOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['approveOrder'],
  description: "Approve Order. Approver approves an order, those exceeds the account's purchase limit",
  input: '@oracle-cx-commerce/actions/order-approval/schema/approve-order/input.json'
};

export const rejectOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['rejectOrder'],
  description: "Reject Order. Approver rejects an order, those exceeds the account's purchase limit",
  input: '@oracle-cx-commerce/actions/order-approval/schema/reject-order/input.json'
};

export const checkOrderRequiresApproval = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['checkOrderRequiresApproval'],
  description: "Check Order Requires Approval. Order's those exceeds the account's purchase limit requires approval",
  input: '@oracle-cx-commerce/actions/order-approval/schema/check-order-requires-approval/input.json'
};
