// @oracle-cx-commerce/react-widgets/profile/profile-create-purchase-list/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-create-purchase-list/config';
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

const widgetResourceKeys = [
  'textPurchaseListItems',
  'headingCreatePurchaseList',
  'textAllFieldsRequired',
  'labelListName',
  'labelDescriptionOptional',
  'headingAddAProduct',
  'labelSearchProductsFilter',
  'actionSave',
  'actionCancel',
  'alertPurchaseListCreated',
  'actionClear',
  'labelProductQuantity',
  'textItemDetails',
  'actionRemoveItem',
  'labelAddToPurchaseList',
  'textBackOrderable',
  'textInStock',
  'textOutOfStock',
  'textPreOrderable',
  'textSelectOptionOnline',
  'textBackOrderableWithAvailabilityDate',
  'textPreOrderableWithAvailabilityDate'
];

export const ProfileCreatePurchaseList = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  actions: ['createPurchaseList', 'notify', 'typeahead', 'clearTypeahead'],
  pageTypes: ['Create Purchase List']
};
