// @oracle-cx-commerce/endpoints/utils/transform/shopper-context.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {isEmptyObject, isObject} from '@oracle-cx-commerce/utils/generic';

/**
 * Transforms Shopper Context properties
 * priceListGroup in global context to be updated with the defaultPriceListGroup id.
 * OCStateData and externalShopperContext will be placed in session context.
 * *
 * @param {Object} json
 * @param {Object} response
 * @returns {Object} clientRepository object with updated global and session contexts.
 */
export const transformShopperContext = (json, response) => {
  if (!isObject(json) || isEmptyObject(json)) {
    return {};
  }

  const result = {
    clientRepository: {
      context: {
        global: {
          priceListGroup: json.defaultPriceListGroup.id
        },
        session: {
          externalShopperContext: json
        }
      }
    }
  };

  if (isObject(response.headers)) {
    result.clientRepository.context.session.OCStateData = response.headers.get('x-ocstatedata');
  }

  return result;
};
