// @oracle-cx-commerce/react-widgets/profile/profile-wish-lists/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-wish-lists/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import {widgetResourceKeys as moveToWishListResourceKeys} from '@oracle-cx-commerce/react-widgets/product/product-add-to-wishlist/meta';

const widgetResourceKeys = [
  'textPurchaseListItems',
  'headingMyLists',
  'labelLoadMore',
  'textRetrievingWishLists',
  'textAllSharedLists',
  'textPurchaseListOwnerDescription',
  'textOwnerLastNameAsc',
  'textOwnerLastNameDesc',
  'textOwnerEmailAsc',
  'textOwnerEmailDesc',
  'textPrivate',
  'textGroup',
  'textShared',
  'actionCreateNewList',
  'headingMyFriendsLists',
  'headingWishLists',
  'actionConfirm',
  'actionCancel',
  'textDeleteList',
  'textConfirmWishListDeletionMessage',
  'closeLinkAltText',
  'alertWishListDeleted',
  'alertWishListCreated',
  'labelListName',
  'alertDuplicateWishList',
  'actionSave',
  'wishlistAccessLevelPrivateHelpText',
  'wishlistAccessLevelGroupHelpText',
  'wishlistAccessLevelSharedHelpText',
  'headingWishlistPrivacySettings',
  'actionWishLists',
  'textImageNotAvailable',
  'textWishListMember',
  'textWishListOwnerDescription',
  'textWishListOwnerEmailDescription',
  'textLeaveList',
  'textConfirmLeaveWishListMessage',
  'alertLeaveWishList',
  'textNoAvailableWishLists',
  ...moveToWishListResourceKeys
];

export const ProfileWishLists = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  actions: ['listWishlists', 'notify', 'notifyClearAll', 'deleteWishlist', 'leaveWishlist', 'getWishlist'],
  pageTypes: ['Wish Lists']
};
