// @oracle-cx-commerce/actions/registration-request/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createOrganizationRegistrationRequest = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
export const createContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
export const listContactRegistrationRequests = () => import('@oracle-cx-commerce/actions/registration-request/actions');
export const getContactRegistrationRequest = () => import('@oracle-cx-commerce/actions/registration-request/actions');
export const updateContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
export const approveContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
export const rejectContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
export const createMultipleContactRegistrationRequests = () =>
  import('@oracle-cx-commerce/actions/registration-request/actions');
