// @oracle-cx-commerce/actions/registration-request/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createOrganizationRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createOrganizationRegistrationRequest'],
  description: 'Create Organization Request. Request for the new Organization registration is created',
  input: '@oracle-cx-commerce/actions/registration-request/schema/create-organization-registration-request/input.json'
};

export const createContactRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createContactRegistrationRequest'],
  description: 'Create Contact Request. Request for the new contact registration is created',
  input: '@oracle-cx-commerce/actions/registration-request/schema/create-contact-registration-request/input.json'
};

export const listContactRegistrationRequests = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listContactRegistrationRequests'],
  description: 'List Contact Requests. This operation is used to get the collection of contact requests.',
  input: '@oracle-cx-commerce/actions/registration-request/schema/list-contact-registration-requests/input.json'
};

export const getContactRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getContactRegistrationRequest'],
  description: 'Get Contact Request. This operation is used to get the Contact request information for the given id.',
  input: '@oracle-cx-commerce/actions/registration-request/schema/get-contact-registration-request/input.json'
};

export const updateContactRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateContactRegistrationRequest'],
  description:
    'Update Contact Request. This operation is used to update the Contact request information, status for the given id.',
  input: '@oracle-cx-commerce/actions/registration-request/schema/update-contact-registration-request/input.json'
};

export const approveContactRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['approveContactRegistrationRequest'],
  description: 'Approve Contact Request. This operation is used to approve profile request.',
  input: '@oracle-cx-commerce/actions/registration-request/schema/approve-contact-registration-request/input.json'
};

export const rejectContactRegistrationRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['rejectContactRegistrationRequest'],
  description: 'Reject Contact Request. This operation is used to reject profile request.',
  input: '@oracle-cx-commerce/actions/registration-request/schema/reject-contact-registration-request/input.json'
};

export const createMultipleContactRegistrationRequests = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createMultipleContactRegistrationRequests'],
  description: 'Create Multiple Contact Requests. Request for the multiple contact registration is created',
  input:
    '@oracle-cx-commerce/actions/registration-request/schema/create-multiple-contact-registration-requests/input.json'
};
