// @oracle-cx-commerce/react-widgets/profile/pending-approval-orders/meta.js
/*
 ** Copyright (c) 2019 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/profile/pending-approval-orders/config';

const widgetResourceKeys = [
  'labelSubmittedBy',
  'labelOrderDate',
  'headingOrdersPendingApproval',
  'labelLoadMore',
  'labelOrderNumber',
  'messageNoOrdersAvailable',
  'textItemsPerPage',
  'textcreationDateasc',
  'textcreationDatedesc',
  'textFilterByOrderNumber',
  'labelAriaSelectItemsPerPage',
  'searchIconAltText',
  'textRetrievingOrdersPendingApproval',
  'textScheduledOrder',
  'labelAriaSortBy',
  'textOrderSearchResult',
  'labelOrderTotal'
];

export const PendingApprovalOrders = {
  packageId: '@oracle-cx-commerce/react-widgets',
  config: mergeDefaultConfig(config),
  resources: buildResources(resourceBundle, widgetResourceKeys),
  availableToAllPages: false,
  pageTypes: ['profile'],
  actions: ['getAllOrdersForProfile', 'notify']
};
