// @oracle-cx-commerce/cli/templates/selectors/template-files/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
const {capitalizeFirstLetter} = require('@oracle-cx-commerce/cli/templates/utils');

function repositoryBody(repository, name) {
  if (name !== undefined) {
    return `/**
 * Selector to extract repositories from a state object.
 */
export const get${capitalizeFirstLetter(name)} = state => state.${repository} || {};
`;
  }

  return `/**
 * Selector to extract repositories from a state object.
 */
export const get${capitalizeFirstLetter(repository)} = state => state.${repository} || {};
`;
}

function tableBody(repository, table, name) {
  if (name !== undefined) {
    return `/**
 * Selector to extract repositories.tables from a state object.
 */
export const get${capitalizeFirstLetter(name)} = state => state.${repository}.${table} || {};
`;
  }

  return `/**
 * Selector to extract repositories.tables from a state object.
 */
export const get${capitalizeFirstLetter(table)} = state => state.${repository}.${table} || {};
`;
}

function entityIdBody(repository, table, entityId, name) {
  if (name !== undefined) {
    return `/**
 * Selector to extract repositories.tables.entityIds from a state object.
 */
export const get${capitalizeFirstLetter(
      name
    )} = (state, {${entityId}}) => state.${repository}.${table}[${entityId}] || {};
`;
  }

  return `/**
 * Selector to extract repositories.tables.entityIds from a state object.
 */
export const get${capitalizeFirstLetter(
    entityId
  )} = (state, {${entityId}}) => state.${repository}.${table}[${entityId}] || {};
`;
}

function entityBody(repository, table, entity, name) {
  if (name !== undefined) {
    return `/**
 * Selector to extract repositories.tables.entities from a state object.
 */
export const get${capitalizeFirstLetter(name)} = state => state.${repository}.${table}.${entity} || {};
`;
  }

  return `/**
 * Selector to extract repositories.tables.entities from a state object.
 */
export const get${capitalizeFirstLetter(entity)} = state => state.${repository}.${table}.${entity} || {};
`;
}

module.exports = options => {
  let content = '';
  if (
    options.repository !== undefined &&
    options.table === undefined &&
    options.entityId === undefined &&
    options.entity === undefined
  ) {
    content += repositoryBody(options.repository, options.name);
  }
  if (
    options.repository !== undefined &&
    options.table !== undefined &&
    options.entityId === undefined &&
    options.entity === undefined
  ) {
    content += tableBody(options.repository, options.table, options.name);
  }
  if (
    options.repository !== undefined &&
    options.table !== undefined &&
    options.entityId !== undefined &&
    options.entity === undefined
  ) {
    content += entityIdBody(options.repository, options.table, options.entityId, options.name);
  }
  if (
    options.repository !== undefined &&
    options.table !== undefined &&
    options.entityId === undefined &&
    options.entity !== undefined
  ) {
    content += entityBody(options.repository, options.table, options.entity, options.name);
  }

  return [
    {
      fileName: 'index.js',
      relativePath: '',
      content
    }
  ];
};
