// @oracle-cx-commerce/react-widgets/checkout/checkout-registration/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'headingCheckoutRegistration',
  'textCreateAnAccountHelper',
  'alertAccountCreated',
  'headingCreateAnAccount',
  'actionCreateAnAccount',
  'alertCreateProfileEmailSentSuccessful',
  'alertCreateProfileSuccessful',
  'labelEmailAddress',
  'labelEmailUpdates',
  'labelFirstName',
  'labelGdprConsentGranted',
  'labelLastName',
  'textLogIn',
  'labelNo',
  'labelPassword',
  'labelPasswordConfirm',
  'alertPasswordNotMatched',
  'labelYes'
];

export const CheckoutRegistration = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: defaultConfig,
  pageTypes: ['checkout-order-confirmation'],
  requiresContext: ['container_context']
};
