// @oracle-cx-commerce/react-widgets/profile/profile-purchase-list-information/components/quick-view/index.jsx
/*
 ** Copyright (c) 2021 Oracle and/or its affiliates.
 */

import {ContainerContext, ProductContext} from '@oracle-cx-commerce/react-ui/contexts';
import React, {useCallback, useState} from 'react';

import Modal from '@oracle-cx-commerce/react-components/modal';
import PageLoader from '@oracle-cx-commerce/react-components/page-loader';
import ProductImage from '@oracle-cx-commerce/react-widgets/product-listing/product-result-image-quick-view/components/product-image';
import ProductInventoryStatus from '@oracle-cx-commerce/react-widgets/product/product-inventory-status/component';
import ProductName from '@oracle-cx-commerce/react-widgets/product/product-name/component';
import ProductPrice from '@oracle-cx-commerce/react-widgets/product/product-price/component';
import ProductVariantOptions from '@oracle-cx-commerce/react-widgets/product/product-variant-options/component';
import PropTypes from 'prop-types';
import QuickViewPlaceholder from '@oracle-cx-commerce/react-widgets/product-listing/product-result-image-quick-view/components/quick-view-placeholder';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/product-listing/product-result-image-quick-view/components/quick-view/styles.css';
import {isEmptyObject} from '@oracle-cx-commerce/utils/generic';

/**
 * Displays the quick view pop up
 *
 * @param {function} props.closeQuickView function that handles closing the pop up
 * @param {boolean} props.showQuickView whether the Quick View pop up is displayed or not
 * @param {string} props.imageUrl the url of the product image in the plp
 * @param {string} props.altText the altText associated with the product image in the plp
 * @param {object} props.product the product details to be displayed in quick view
 */
const QuickView = props => {
  const {closeQuickView, showQuickView, imageUrl, altText, product, record, addToList, labelAddToButton} = props;

  const {variantOptionsSeed, variantToSkuLookup = {}, childSKUs = []} = product;

  const skuNotInProducts = (childSKUs, skuId) => {
    const sku = childSKUs.find(chidSKU => chidSKU === skuId);

    return sku === undefined ? true : false;
  };

  const initialState = {
    skuId: null,
    variantOptions: {},
    qty: 1
  };

  const [selections, setSelections] = useState(initialState);

  // For products with no variant options, update the skuId to retrieve stock status
  if (
    isEmptyObject(selections) ||
    (isEmptyObject(variantOptionsSeed) && selections.skuId !== variantToSkuLookup['']) ||
    (!isEmptyObject(variantOptionsSeed) && selections.skuId && skuNotInProducts(childSKUs, selections.skuId))
  ) {
    setSelections({...selections, skuId: variantToSkuLookup['']});
  }

  /**
   * Handler for closing the Quick View modal
   */
  const onClose = useCallback(() => {
    // Reset to default state of product detail selections
    setSelections(initialState);
    // close quick view modal
    closeQuickView();
  }, [setSelections, initialState, closeQuickView]);

  /**
   * Handler for AddToCartButton
   * Dispatches action when add to cart is clicked
   * Closes quick view on success
   * @param  {Object} Object
   */
  const onAddToList = useCallback(() => {
    const variantData = {};
    const selectedVariants = [];
    Object.values(variantOptionsSeed).forEach(item => {
      const variant = {};
      variant.optionName = item.optionName;

      Object.values(item.options).forEach(option => {
        if (option.value === item.selectedValue) {
          variant.optionValue = option.name;
        }
      });
      selectedVariants.push(variant);
    });
    variantData.selectedVariants = selectedVariants;
    variantData.skuId = selections.skuId;
    addToList(record, variantData);
    // close quick view modal
    closeQuickView();
  }, [closeQuickView, record, addToList, selections, variantOptionsSeed]);

  return (
    <Styled id="QuickView" css={css}>
      <Modal show={showQuickView} onClose={onClose}>
        <div className="QuickView">
          <PageLoader show={isEmptyObject(product)}>
            <QuickViewPlaceholder />
          </PageLoader>
          <ProductContext.Provider value={product}>
            <ContainerContext.Provider value={{selections, setSelections}}>
              <ProductImage imageUrl={imageUrl} altText={altText} />
              <div className="QuickView__ProductDetails">
                <ProductName />
                <ProductPrice {...props} />
                <ProductVariantOptions {...props} />
                <ProductInventoryStatus {...props} />
                <button data-testid="Add-To-PurchaseList-Button" type="submit" onClick={onAddToList}>
                  {labelAddToButton}
                </button>
              </div>
            </ContainerContext.Provider>
          </ProductContext.Provider>
        </div>
      </Modal>
    </Styled>
  );
};

QuickView.propTypes = {
  closeQuickView: PropTypes.func.isRequired,
  showQuickView: PropTypes.bool.isRequired,
  imageUrl: PropTypes.string.isRequired,
  altText: PropTypes.string.isRequired,
  product: PropTypes.shape({
    variantOptionsSeed: PropTypes.shape({}),
    variantToSkuLookup: PropTypes.shape({})
  }).isRequired
};

export default React.memo(QuickView);
