// @oracle-cx-commerce/endpoints/organization-member/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const listOrganizationMembers = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/list-organization-members');
export const getOrganizationMember = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/get-organization-member');
export const updateOrganizationMember = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/update-organization-member');
export const createOrganizationMember = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/create-organization-member');
export const removeOrganizationMembers = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/remove-organization-members');
export const removeOrganizationMemberRoles = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/remove-organization-member-roles');
export const addOrganizationMemberRoles = () =>
  import('@oracle-cx-commerce/endpoints/organization-member/add-organization-member-roles');
