// @oracle-cx-commerce/endpoints/organization-member/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export {default as listOrganizationMembers} from '@oracle-cx-commerce/endpoints/organization-member/list-organization-members/meta';
export {default as getOrganizationMember} from '@oracle-cx-commerce/endpoints/organization-member/get-organization-member/meta';
export {default as updateOrganizationMember} from '@oracle-cx-commerce/endpoints/organization-member/update-organization-member/meta';
export {default as createOrganizationMember} from '@oracle-cx-commerce/endpoints/organization-member/create-organization-member/meta';
export {default as removeOrganizationMembers} from '@oracle-cx-commerce/endpoints/organization-member/remove-organization-members/meta';
export {default as removeOrganizationMemberRoles} from '@oracle-cx-commerce/endpoints/organization-member/remove-organization-member-roles/meta';
export {default as addOrganizationMemberRoles} from '@oracle-cx-commerce/endpoints/organization-member/add-organization-member-roles/meta';
