// @oracle-cx-commerce/cli/templates/widget/template-files/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

const configImport = `import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import * as resources from '../../../core/ui/locales';`;

const configLocalesKeys = `const resourceKeys = ['includeExtraInfoLabel', 'includeExtraInfoHelpText'];`;

function configBody() {
  return `${configImport}

/**
 * The list below declares which of the locale resources should be made available to the widget
 */
${configLocalesKeys}

/**
 * Design Studio configuration properties for the WidgetName component.
 */
const config = {
  properties: [
    {
      id: 'configIncludeExtraInfo',
      /**
       * Supported config types:
       *  - stringType
       *  - booleanType
       *  - optionType
       *  - multiSelectOptionType
       *  - sectionTitleType
       *  - collectionType
       *  - mediaType
       */
      type: 'booleanType',
      defaultValue: false,
      labelResourceId: 'includeExtraInfoLabel',
      helpTextResourceId: 'includeExtraInfoHelpText'
    }
  ],
  locales: buildConfigResources(resources, resourceKeys)
};

export default config;
`;
}

// eslint-disable-next-line no-unused-vars
module.exports = options => {
  const content = configBody();

  return [
    {
      fileName: 'config.js',
      relativePath: '',
      content
    }
  ];
};
