// @oracle-cx-commerce/actions/organization-member/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const listOrganizationMembers = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listOrganizationMembers'],
  description:
    'This operation is used to get the collection of organization member profiles from Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-member/schema/list-organization-members/input.json'
};

export const getOrganizationMember = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrganizationMember'],
  description:
    'This operation is used to get the profile information of a organization member that is associated to an account from Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/organization-member/schema/get-organization-member/input.json'
};

export const updateOrganizationMember = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateOrganizationMember'],
  description:
    'This operation is used to update customer name, email, role, status of a customer profile and contact dynamic properties.',
  input: '@oracle-cx-commerce/actions/organization-member/schema/update-organization-member/input.json'
};

export const createOrganizationMember = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createOrganizationMember'],
  description: 'This operation is used to create a profile in Oracle Commerce Cloud for a customer.',
  input: '@oracle-cx-commerce/actions/organization-member/schema/create-organization-member/input.json'
};

export const removeOrganizationMembers = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['removeOrganizationMembers'],
  description: 'This operation is used to remove members from the organization by delegatedAdmin.',
  input: '@oracle-cx-commerce/actions/organization-member/schema/remove-organization-members/input.json'
};

export const addOrganizationMemberRoles = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addOrganizationMemberRoles'],
  description: 'Add Member Roles. This operation is used to assign organization based or global roles to a user',
  input: '@oracle-cx-commerce/actions/organization-member/schema/add-organization-member-role/input.json'
};

export const removeOrganizationMemberRoles = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['removeOrganizationMemberRoles'],
  description: 'Remove Member Roles. This operation is used to remove organization based or global roles from a user',
  input: '@oracle-cx-commerce/actions/organization-member/schema/remove-organization-member-role/input.json'
};
