// @oracle-cx-commerce/react-widgets/common/quick-order/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';

const configResourceKeys = [
  'skuEntryFieldsLabel',
  'skuEntryFieldsHelpText',
  'skuSuggestionsLabel',
  'skuSuggestionsHelpText',
  'maxImportRowsLabel',
  'maxImportRowsHelpText'
];

const quickOrderConfiguration = {
  properties: [
    {
      id: 'skuEntryFields',
      type: 'stringType',
      labelResourceId: 'skuEntryFieldsLabel',
      defaultValue: '5',
      required: true,
      helpTextResourceId: 'skuEntryFieldsHelpText'
    },
    {
      id: 'skuSuggestions',
      type: 'stringType',
      labelResourceId: 'skuSuggestionsLabel',
      defaultValue: '5',
      required: true,
      helpTextResourceId: 'skuSuggestionsHelpText'
    },
    {
      id: 'maxImportRows',
      type: 'stringType',
      labelResourceId: 'maxImportRowsLabel',
      defaultValue: '100',
      required: true,
      helpTextResourceId: 'maxImportRowsHelpText'
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
};

export default quickOrderConfiguration;
