// @oracle-cx-commerce/actions/order/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const getOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrder'],
  input: '@oracle-cx-commerce/actions/order/schema/get-order/input.json',
  description:
    "Get Order. Get an order. Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.This endpoint returns the order details irrespective of organization ID passed in request header. If user doesn't belong to order's organization, then the user has no access to order."
};

export const cancelOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['cancelOrder'],
  input: '@oracle-cx-commerce/actions/order/schema/cancel-order/input.json',
  description:
    'Cancel Order. Cancels the order with the specified id. Requires the order to be within the given remorse period and for the current logged in user to be the owner of the order.'
};

export const getOrderConfirmation = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrderConfirmation'],
  description:
    "Get Order Confirmation. Get an order confirmation details. Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.This endpoint returns the order details irrespective of organization ID passed in request header. If user doesn't belong to order's organization, then the user has no access to order.",
  input: '@oracle-cx-commerce/actions/order/schema/get-order-confirmation/input.json'
};

export const getAllOrdersForProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getAllOrdersForProfile'],
  description:
    'Get All Orders For Profile. Retrieves all orders associated with the logged in user. Also if the user is approver, orders of other users of the organziation which are requiring approval will be sent, Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.',
  input: '@oracle-cx-commerce/actions/order/schema/get-all-orders-for-profile/input.json'
};

export const getValidActionsForOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getValidActionsForOrder'],
  description: 'Get the valid actions for a given order.',
  input: '@oracle-cx-commerce/actions/order/schema/get-valid-actions-for-order/input.json'
};

export const initiateCancelOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['initiateCancelOrder'],
  description:
    'Initiates cancellation of in-flight order with the specified ID. Requires the order to be within the given extended remorse period and for the currently logged-in user to be the one who created the order.',
  input: '@oracle-cx-commerce/actions/order/schema/initiate-cancel-order/input.json'
};

export const submitCancelOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['submitCancelOrder'],
  description:
    'Submits the cancel order with the specified ID to fulfillment. The cancel order object is created during the initiateCancel operation on the original order.',
  input: '@oracle-cx-commerce/actions/order/schema/submitCancelOrder/input.json'
};

export const copyOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['copyOrder'],
  description: 'Duplicates an order with the specified ID into a new incomplete order',
  input: '@oracle-cx-commerce/actions/order/schema/copy-order/input.json'
};
