// @oracle-cx-commerce/react-widgets/common/promotion-upsell-messaging/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

export const promotionUpsellConfigResourceKeys = [
  'configPromotionUpsellTagsHelpText',
  'configPromotionUpsellTagsLabel',
  'configPromotionUpsellMessageLimitHelpText',
  'configPromotionUpsellMessageLimitLabel'
];
export const promotionUpsellConfigProperties = [
  {
    id: 'tags',
    type: 'stringType',
    name: 'tags',
    helpTextResourceId: 'configPromotionUpsellTagsHelpText',
    labelResourceId: 'configPromotionUpsellTagsLabel',
    defaultValue: ''
  },
  {
    id: 'messageLimit',
    type: 'stringType',
    name: 'messageLimit',
    helpTextResourceId: 'configPromotionUpsellMessageLimitHelpText',
    labelResourceId: 'configPromotionUpsellMessageLimitLabel',
    defaultValue: '1',
    required: true
  }
];

const promotionUpsellConfig = mergeDefaultConfig({
  properties: [...promotionUpsellConfigProperties],
  locales: buildConfigResources(resourceBundle, promotionUpsellConfigResourceKeys)
});

export default promotionUpsellConfig;
