// @oracle-cx-commerce/react-widgets/cart/cart-shipping-estimate/meta.js
/*
 ** Copyright (c) 2021 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import cartShippingEstimateConfig from '@oracle-cx-commerce/react-widgets/cart/cart-shipping-estimate/config';

export const cartShippingEstimateWidgetResourceKeys = [
  'alertShippingGroupUpdated',
  'headingShippingCalculator',
  'headingAvailableShippingOptions',
  'textNoShippingMethods',
  'textEnterAddressForShippingEstimate',
  'textCartContainsMultipleShippingGroups',
  'textInvalidField',
  'textRequiredField',
  'labelCountry',
  'labelEstimateShipping',
  'labelZipCode',
  'labelState',
  'labelStreetAddress',
  'labelTownCity'
];
export const CartShippingEstimate = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, cartShippingEstimateWidgetResourceKeys),
  config: cartShippingEstimateConfig,
  availableToAllPages: false,
  pageTypes: ['cart'],
  requiresContext: ['cart_context']
};
