// @oracle-cx-commerce/react-widgets/checkout/checkout-multi-shipping-details/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/checkout/checkout-multi-shipping-details/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'actionConfirm',
  'labelCancel',
  'headingShipEverythingToOneAddress',
  'actionCancel',
  'labelShippingOptions',
  'headingShippingOptions',
  'textEnterAShippingAddress',
  'textNoShippingMethods',
  'headingHomeDelivery',
  'headingShippingTo',
  'labelEditAddress',
  'labelOpenAddressBook',
  'textOpenAddressBookAndChooseAddress',
  'labelAddAnAddress',
  'labelAddANewAddress',
  'actionMoveToWishList',
  'messagePartialBackOrder',
  'messageInsufficientStock',
  'messageInsufficientStockAtStore',
  'messageItemNoLongerAvailable',
  'messageQuantityManditory',
  'labelProductQuantity',
  'actionRemoveItem',
  'shippingSurchargeText',
  'textTotal',
  'textItemDetails',
  'textItemPrice',
  'textQuantity',
  'textGiftItem',
  'textSelectGiftMessage',
  'actionSelect',
  'textFreeProduct',
  'textFreeGift',
  'textFree',
  'textChange',
  'textSelectGift',
  'alertAddToCartAdding',
  'textInStock',
  'textOutOfStock',
  'textPreOrderable',
  'textBackOrderable',
  'textSelectOptionOnline',
  'actionAddToCartPreOrder',
  'alertTotalItemQuantityExceeded',
  'textGWPInvalidation',
  'alertOutOfStock',
  'alertCartHeading',
  'alertNoLongerForSale',
  'messageAtTheRate',
  'messagePriceChange',
  'alertPriceIncreased',
  'alertPriceDecreased',
  'messagePartialPreOrder',
  'messageStatusPreOrder',
  'messageStatusBackOrder',
  'closeLinkAltText',
  'textAllFieldsRequired',
  'textInvalidField',
  'textRequiredField',
  'labelNickName',
  'labelCompanyName',
  'labelFirstName',
  'labelLastName',
  'labelCountry',
  'labelZipCode',
  'labelState',
  'labelStreetAddress',
  'labelTownCity',
  'labelPhoneNumberOptional',
  'labelSaveAsANewProfileAddress',
  'labelSaveAsANewAccountAddress',
  'labelCancel',
  'buttonSaveAndContinue',
  'textAccountAddressBook',
  'textAccountAddress',
  'textProfileAddress',
  'textDefaultAddresses',
  'textProfileAddresses',
  'textInheritedAddresses',
  'textAddressesForThisAccount',
  'labelDefaultShippingAddress',
  'labelDefaultBillingAddress',
  'labelNoDefaultShippingAddressAvailable',
  'labelNoDefaultBillingAddressAvailable',
  'labelNoDefaultAddressesAvailable',
  'labelNoProfileAddressesAvailable',
  'labelNoAccountAddressesAvailable',
  'labelNoInheritedAddressesAvailable',
  'textLoading',
  'textLoadMoreAccountAddress',
  'textLoadMoreProfileAddress',
  'textLoadMoreInheritedAddress',
  'textUseThisAddress',
  'headingAddressBook',
  'labelAddressBook',
  'labelDefaultAddress',
  'headingAddShippingAddress',
  'headingEditShippingAddress',
  'alertShippingGroupUpdated',
  'alertAddressCreatedSuccessfully',
  'alertAddressChangedSuccessfully',
  'headingDeliveryAddress',
  'labelSelectAnAddress',
  'actionShipThisItemToMultipleAddresses',
  'actionShipAllItemsToASingleAddress',
  'actionShipItemsToMultipleAddresses',
  'labelEdit',
  'headingItemsWillBeShippedTogether',
  'alertOrderWillShipToOneAddress',
  'alertConfirmShipToOneAddress',
  'alertNoLongerForSaleQuote',
  'alertOrderHeading',
  'alertOutOfStockQuote'
];

export const CheckoutMultiShippingDetails = {
  availableToAllPages: false,
  config: mergeDefaultConfig(config),
  packageId: '@oracle-cx-commerce/react-widgets',
  pageTypes: ['checkout-shipping'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  requiresContext: ['checkout_shipping_context', 'cart_context']
};
