// @oracle-cx-commerce/react-widgets/checkout/checkout-scheduled-order/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'labelScheduleOrder',
  'labelScheduleOrderSubscript',
  'labelScheduleOrderPaymentSubscript',
  'labelScheduleThisOrder',
  'labelScheduledOrderName',
  'labelStartDate',
  'labelEndDate',
  'labelFrequency',
  'labelSuspendOrder',
  'labelSuspendScheduledOrder',
  'labelSelectDaysOfWeek',
  'labelSundayAbbreviation',
  'labelMondayAbbreviation',
  'labelTuesdayAbbreviation',
  'labelWednesdayAbbreviation',
  'labelThursdayAbbreviation',
  'labelFridayAbbreviation',
  'labelSaturdayAbbreviation',
  'labelSelectWeeksOfMonth',
  'labelFirstWeek',
  'labelSecondWeek',
  'labelThirdWeek',
  'labelFourthWeek',
  'labelLastWeek',
  'labelDaily',
  'labelWeekly',
  'labelMonthly',
  'labelOnceADay',
  'labelOnceAMonth',
  'labelEveryTwoMonths',
  'labelQuarterly',
  'textDaysMandatory',
  'textWeeksMandatory',
  'textScheduleDateInvalid',
  'textNameMandatory',
  'textStartDateMandatory',
  'textEndDateInvalid',
  'textFrequencyMandatory',
  'labelAriaSelectFrequency',
  'textEndDateFormatMandatory',
  'textStartDateInvalid'
];

export const CheckoutScheduledOrder = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: defaultConfig,
  availableToAllPages: false,
  pageTypes: ['checkout-shipping', 'Scheduled Order Details']
};
