// @oracle-cx-commerce/endpoints/payment/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addPayments = () => import('@oracle-cx-commerce/endpoints/payment/add-payments');
export const applyPayments = () => import('@oracle-cx-commerce/endpoints/payment/apply-payments');
export const calculateRemainingBalance = () =>
  import('@oracle-cx-commerce/endpoints/payment/calculate-remaining-balance');
export const deleteAppliedPayment = () => import('@oracle-cx-commerce/endpoints/payment/delete-applied-payment');
export const getBillingCountries = () => import('@oracle-cx-commerce/endpoints/payment/get-billing-countries');
export const getCardTypes = () => import('@oracle-cx-commerce/endpoints/payment/get-card-types');
export const getPaymentGroup = () => import('@oracle-cx-commerce/endpoints/payment/get-payment-group');
export const getPaymentInitiatedOrder = () =>
  import('@oracle-cx-commerce/endpoints/payment/get-payment-initiated-order');
export const getPaymentConfigurations = () =>
  import('@oracle-cx-commerce/endpoints/payment/get-payment-configurations');
export const initiatePayment = () => import('@oracle-cx-commerce/endpoints/payment/initiate-payment');
export const inquireBalance = () => import('@oracle-cx-commerce/endpoints/payment/inquire-balance');
export const updateAppliedPayment = () => import('@oracle-cx-commerce/endpoints/payment/update-applied-payment');
