// @oracle-cx-commerce/react-widgets/profile/profile-purchase-list-information/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-purchase-list-information/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'labelSearchProductsFilter',
  'headingAddAProduct',
  'labelCancel',
  'actionSelectAll',
  'actionDeselectAll',
  'labelAddSelectedItemsToCart',
  'labelSaveChangesToItemsInPurchaseList',
  'labelProductQuantity',
  'textItemDetails',
  'actionRemoveItem',
  'alertMergeItemsToCart',
  'alertMergeItemsToCartConfirm',
  'textShoppingCartNotEmpty',
  'actionConfirm',
  'actionCancel',
  'closeLinkAltText',
  'alertPurchaseListUpdated',
  'actionClear',
  'labelAddToPurchaseList',
  'textBackOrderable',
  'textInStock',
  'textOutOfStock',
  'textPreOrderable',
  'textSelectOptionOnline',
  'textBackOrderableWithAvailabilityDate',
  'textPreOrderableWithAvailabilityDate',
  'alertCannotFetchSku',
  'textNoItemsToDisplay',
  'textItemisNotAvailable',
  'textSku'
];
export const ProfilePurchaseListInformation = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  pageTypes: ['profile'],
  requiresContext: ['container_context'],
  actions: [
    'clearTypeahead',
    'listSkus',
    'updatePurchaseList',
    'notify',
    'typeahead',
    'qaddItemsToCart',
    'checkPurchaseListEditAccess'
  ]
};
