// @oracle-cx-commerce/endpoints/registration-request/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createOrganizationRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/create-organization-registration-request');
export const createContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/create-contact-registration-request');

export const listContactRegistrationRequests = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/list-contact-registration-requests');
export const getContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/get-contact-registration-request');
export const updateContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/update-contact-registration-request');
export const approveContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/approve-contact-registration-request');
export const rejectContactRegistrationRequest = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/reject-contact-registration-request');
export const createMultipleContactRegistrationRequests = () =>
  import('@oracle-cx-commerce/endpoints/registration-request/create-multiple-contact-registration-requests');
