// @oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/multi-select-menu/multi-select-checkbox-horizontal.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import React, {useEffect} from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/multi-select-menu/styles.css';
import PropTypes from 'prop-types';

/**
 * This component will render a multi select menu cards UI control.
 *
 * @param {Object} props.inputRef
 *   A reference to the DOM element created in the render method. This will be
 *   passed from the parent component.
 * @param {String} props.uniqueId
 *   A unique id for the rendered DOM element, it should be unique within the HTML
 *   document. This will be passed from the parent component.
 * @param {String} props.name
 *   A display name (label) for the form control. This will be passed from the
 *   parent component.
 * @param {String} props.variableName
 *   The variable name used to connnect a form control to its corresponding value in
 *   configData . This will be passed from the parent component.
 * @param {String} props.defaultValue
 *   The initial value for the form control. It is retrieved from the configData
 *   object from the configurations repository store. This will be passed from
 *   the parent component. See:
 *   https://reactjs.org/docs/uncontrolled-components.html#default-values
 * @param {Function} props.handleChange
 *   The function to be called when a user has changed the value of the form control.
 *   This will be passed from the parent component.
 * @param {Boolean} props.updatable
 *   If false the form control should be disabled / read only. This will be
 *   passed from the parent component.
 * @param {Boolean} props.required
 *   If true the form control should be a mandatory / required field. This will be
 *   passed from the parent component.
 * @param {Array} props.menuItems
 *   The collection of menu items to be displayed by the form control. This will be
 *   passed from the parent component
 */
const MultiSelectCheckboxHorizontal = ({
  inputRef,
  uniqueId,
  name,
  variableName,
  defaultValue,
  handleChange,
  updatable,
  required,
  menuItems
}) => {
  const multiSelectInputRefs = [];

  const setMultiSelectInputRef = ref => {
    multiSelectInputRefs.push(ref);
  };

  // this is required for interact
  // inputRef for this component will be a field set
  useEffect(() => {
    multiSelectInputRefs.forEach(msInputRef => {
      msInputRef.checked = defaultValue.includes(msInputRef.value);
    });
  }, [defaultValue, multiSelectInputRefs]);

  /**
   *
   */
  const handleMultiSelectMenuChange = changeEvent => {
    const {value: newValue, checked} = changeEvent.target;
    const selectedOptions = defaultValue
      .filter(item => item !== newValue)
      .map(currentValue => {
        return {value: currentValue};
      });

    if (checked) {
      selectedOptions.push({value: newValue});
    }

    handleChange({items: selectedOptions});
  };

  return (
    <Styled id="ConfiguratorMultiSelectCheckboxHorizontal" css={css}>
      <div className="ConfiguratorMultiSelectCheckboxHorizontal__Container">
        <fieldset className="ConfiguratorMultiSelectCheckboxHorizontal__Fieldset" ref={inputRef}>
          <legend className="ConfiguratorMultiSelectCheckboxHorizontal__Legend">{name}</legend>
          {menuItems.map((item, index) => {
            return (
              <label htmlFor={`${uniqueId}-${index}`} key={item.value}>
                <input
                  id={`${uniqueId}-${index}`}
                  type="checkbox"
                  value={item.value}
                  onChange={handleMultiSelectMenuChange}
                  name={variableName}
                  defaultChecked={defaultValue.includes(item.value)}
                  required={required}
                  disabled={!updatable}
                  ref={setMultiSelectInputRef}
                />
                {item.displayValue}
              </label>
            );
          })}
        </fieldset>
      </div>
    </Styled>
  );
};

MultiSelectCheckboxHorizontal.propTypes = {
  /**
   * A reference to the DOM element created in the render method.
   * This will be passed from the parent component.
   */
  inputRef: PropTypes.shape(Object),
  /**
   * A unique id for the rendered DOM element, it should be unique within the HTML
   * document. This will be passed from the parent component.
   */
  uniqueId: PropTypes.string.isRequired,
  /**
   * A display name (label) for the form control. This will be passed from the
   * parent component.
   */
  name: PropTypes.string.isRequired,
  /**
   * The variable name used to connect a form control to its corresponding value in
   * configData . This will be passed from the parent component.
   */
  variableName: PropTypes.string.isRequired,
  /**
   * The initial value for the form control. It is retrieved from the configData
   * object from the configurations repository store. This will be passed from
   * the parent component. See:
   * https://reactjs.org/docs/uncontrolled-components.html#default-values
   */
  defaultValue: PropTypes.arrayOf(PropTypes.string).isRequired,
  /**
   * The function to be called when a user has changed the value of the form control.
   * This will be passed from the parent component.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * If false the form control should be disabled / read only. This will be
   * passed from the parent component.
   */
  updatable: PropTypes.bool.isRequired,
  /**
   * If true the form control should be a mandatory / required field. This will be
   * passed from the parent component.
   */
  required: PropTypes.bool.isRequired,
  /**
   * The collection of menu items to be displayed by the form control. This will be
   * passed from the parent component
   */
  menuItems: PropTypes.arrayOf(Object).isRequired
};

MultiSelectCheckboxHorizontal.defaultProps = {
  inputRef: null
};

export default MultiSelectCheckboxHorizontal;
