// @oracle-cx-commerce/react-widgets/profile/profile-wish-list-details/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-wish-list-details/config';
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

const widgetResourceKeys = [
  'labelCancel',
  'actionSelectAll',
  'actionDeselectAll',
  'textItemDetails',
  'actionRemoveItem',
  'alertMergeItemsToCart',
  'alertMergeItemsToCartConfirm',
  'textShoppingCartNotEmpty',
  'actionConfirm',
  'actionCancel',
  'closeLinkAltText',
  'actionClear',
  'textAddCommentQuantityAndPriority',
  'textEditCommentQuantityAndPriority',
  'labelEnterComments',
  'labelLow',
  'labelMedium',
  'labelHigh',
  'labelSelectPriority',
  'labelDelete',
  'labelEditWishListProduct',
  'labelMove',
  'textPrivate',
  'textGroup',
  'textShared',
  'actionShare',
  'textEditList',
  'labelViewAllLists',
  'textDeleteList',
  'labelInviteFriends',
  'actionSave',
  'labelQuantityWanted',
  'labelHidewishListComment',
  'labelPriority',
  'headingMoveToAnotherList',
  'textConfirmWishListDeletionMessage',
  'alertWishListDeleted',
  'labelAriaSortText',
  'textRetrievingWishList',
  'labelPrioritySort',
  'labelDateAdded',
  'labelListName',
  'alertDuplicateWishList',
  'wishlistAccessLevelPrivateHelpText',
  'wishlistAccessLevelGroupHelpText',
  'wishlistAccessLevelSharedHelpText',
  'headingWishlistPrivacySettings',
  'alertPriceUnavailable',
  'actionSend',
  'textInviteFriendsMessage',
  'textNotifyMeViaEmail',
  'labelProductAddedToList',
  'labelCommentAddedToList',
  'labelMemberJoinsList',
  'actionRemoveFriend',
  'labelEmailAddress',
  'textAddEmailAddress',
  'actionQuickOrderAddToCart',
  'textWishlistEmailSubject',
  'textWishlistEmailBodyPara1',
  'textWishlistEmailBodyPara2',
  'alertWishListUpdated',
  'alertWishListProductDeleted',
  'textNoAvailableProductWishLists'
];
export const ProfileWishListDetails = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  pageTypes: ['profile'],
  actions: [
    'getWishlist',
    'notify',
    'addItemsToCart',
    'notifyClearAll',
    'addWishlistItemComments',
    'updateWishlistProduct',
    'createInvitation',
    'listWishlists',
    'moveWishlistProduct'
  ]
};
