// @oracle-cx-commerce/endpoints/cart/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const applyCouponsToCart = () => import('@oracle-cx-commerce/endpoints/cart/apply-coupons-to-cart');
export const checkoutCart = () => import('@oracle-cx-commerce/endpoints/cart/checkout-cart');
export const getCart = () => import('@oracle-cx-commerce/endpoints/cart/get-cart');
export const getUpsellMessages = () => import('@oracle-cx-commerce/endpoints/cart/get-upsell-messages');
export const listShippingMethodsForCart = () =>
  import('@oracle-cx-commerce/endpoints/cart/list-shipping-methods-for-cart');
export const priceCart = () => import('@oracle-cx-commerce/endpoints/cart/price-cart');
export const removeCouponsFromCart = () => import('@oracle-cx-commerce/endpoints/cart/remove-coupons-from-cart');
export const updateCart = () => import('@oracle-cx-commerce/endpoints/cart/update-cart');
export const listCarts = () => import('@oracle-cx-commerce/endpoints/cart/list-carts');
export const deleteCart = () => import('@oracle-cx-commerce/endpoints/cart/delete-cart');
export const switchCart = () => import('@oracle-cx-commerce/endpoints/cart/switch-cart');
export const saveCart = () => import('@oracle-cx-commerce/endpoints/cart/save-cart');
