// @oracle-cx-commerce/react-widgets/category/collection-navigation-mobile/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const configResourceKeys = [
  'configNumberOfChildCategoriesToDisplayHelpText',
  'configNumberOfChildCategoriesToDisplayLabel',
  'configCategoryMediaPropertyLabel',
  'configCategoryMediaPropertyHelpText'
];

const config = mergeDefaultConfig({
  properties: [
    {
      id: 'numberOfChildCategoriesToDisplay',
      type: 'stringType',
      name: 'numberOfChildCategoriesToDisplay',
      helpTextResourceId: 'configNumberOfChildCategoriesToDisplayHelpText',
      labelResourceId: 'configNumberOfChildCategoriesToDisplayLabel',
      minLength: 1,
      maxLength: 1,
      defaultValue: '3',
      pattern: '[1-9]',
      required: true
    },
    {
      id: 'categoryMediaProperty',
      type: 'stringType',
      name: 'categoryMediaProperty',
      labelResourceId: 'configCategoryMediaPropertyLabel',
      helpTextResourceId: 'configCategoryMediaPropertyHelpText'
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
});

export default config;
