// @oracle-cx-commerce/react-widgets/category/collection-navigation-mobile/selectors.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import {getGlobalContext, getCategories, getPage} from '@oracle-cx-commerce/commerce-utils/selector';

/**
 * Returns expanded list of root categories
 * @param {Array} [rootCategoryIds=[]] root category ids
 * @param {Object} [catalogCategories={}] all fetched catalog categories
 * @return {Array} root categories
 */
export const getRootCategoriesData = (rootCategoryIds = [], catalogCategories = {}) =>
  rootCategoryIds.map(categoryId => catalogCategories[categoryId]);

/**
 * Returns expanded list of categories
 *
 * @param {Array} [childCategories=[]] child categories
 * @param {Object} [catalogCategories={}] all fetched catalog categories
 * @return {Array} categories
 */
export const getExpandedCategories = (childCategories = [], catalogCategories = {}) =>
  childCategories.map(category => catalogCategories[category.id] || category);

export const getComponentData = state => ({
  rootCategoryIds: getGlobalContext(state).categories || [],
  catalogCategories: getCategories(state),
  currentPage: getPage(state)
});
