// @oracle-cx-commerce/react-widgets/profile/account-address-form/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'headingAddAccountAddress',
  'headingEditAccountAddress',
  'textAllFieldsRequired',
  'textInvalidField',
  'textRequiredField',
  'labelNickName',
  'labelCompanyName',
  'labelCountry',
  'labelZipCode',
  'labelState',
  'labelStreetAddress',
  'labelTownCity',
  'labelPhoneNumberOptional',
  'labelMakeDefaultBillingAddress',
  'labelMakeDefaultShippingAddress',
  'labelCancel',
  'labelSave',
  'alertAddressChangedSuccessfully',
  'alertAddressCreatedSuccessfully'
];
export const AccountAddressForm = {
  availableToAllPages: false,
  config: defaultConfig,
  packageId: '@oracle-cx-commerce/react-widgets',
  pageTypes: ['profile'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  actions: [
    'addOrganizatonAddress',
    'getOrganizationDefaultAddresses',
    'getOrganizationAddress',
    'getProfileAddress',
    'notify',
    'updateOrganizationAddress'
  ],
  fetchers: ['fetchShippingCountries']
};
