// @oracle-cx-commerce/react-widgets/profile/contact-registration-detail/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'labelStatus',
  'labelRequestDate',
  'labelRequestId',
  'labelSite',
  'labelNotesFromTheRequester',
  'labelRejectionComments',
  'textRegistrationRequestsNew',
  'textRegistrationRequestsReview',
  'textMoreInfoRequired',
  'textRegistrationRequestRejected',
  'textNameNotRetained',
  'headingRegistrationRequest',
  'closeRegistrationRequestModalAltText',
  'labelAriaFilterText',
  'textRegistrationRequestAccept',
  'textRegistrationRequestReject',
  'actionCancel',
  'actionSave',
  'actionAccept',
  'actionReject',
  'headingAcceptConfirmation',
  'headingRejectConfirmation',
  'textAcceptRequestDialogDescription',
  'textAcceptDialogPlaceholder',
  'labelCancel',
  'labelOk',
  'textRejectRequestDialogDescription',
  'textRejectDialogPlaceholder',
  'textApprovedRejectRequestCommentsHeader',
  'labelStatusWithoutDelimiter',
  'alertContactRequestUpdated'
];

export const ContactRegistrationDetail = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  pageTypes: ['contact-request'],
  config: defaultConfig
};
