// @oracle-cx-commerce/eslint-plugin-osf/test/unit/~disabled~oracle-copyright.spec~.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

const {RuleTester} = require('eslint');
const rule = require('../../rules/oracle-copyright');

const ruleTester = new RuleTester({parserOptions: {ecmaVersion: 2015, sourceType: 'module'}});
const YEAR = new Date().getFullYear();

/**
 * Disabling until compatibility between ESlint 8 and Jest v26 (v27) is resolved
 * https://github.com/eslint/eslint/issues/14936
 */

describe('Eslint custom OSF plugin tests - Oracle Copyright rule', () => {
  ruleTester.run('custom named export rule', rule, {
    valid: [
      // Typical Happy path cases
      {
        code: `/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        code: `/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        // Unix style line endings
        code: `/*\n ** Copyright (c) 2020 Oracle and/or its affiliates.\n */\n\nlet someStatement;`
      },
      {
        // Windows style line endings
        code: `/*\r\n ** Copyright (c) 2020 Oracle and/or its affiliates.\r\n */\r\n\r\nlet someStatement;`
      },
      {
        // Shebang followed by valid copyright
        code: `#!/usr/bin/env node
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        // Shebang followed by valid copyright with multiple line breaks
        code: `#!/usr/bin/env node

/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

let someStatement;`
      }
    ],
    invalid: [
      {
        // no copyright no comment
        code: `let someStatement;`,
        errors: [
          {
            messageId: 'missingCopyrightMessage',
            type: 'Program'
          }
        ],
        output: `/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        // some copyright year less that 2000
        code: `/*
 ** Copyright (c) 1999 Oracle and/or its affiliates.
 */

let someStatement;`,
        errors: [
          {
            messageId: 'copyrightNeedsFix',
            type: 'Program'
          }
        ],
        output: `/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

/*
 ** Copyright (c) 1999 Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        // first line code followed by some block comment not related to copyright.
        code: `let someStatement;
/*
 ** some block comment not related to copyright.
 */
let someOtherComment`,
        errors: [
          {
            messageId: 'missingCopyrightMessage',
            type: 'Program'
          }
        ],
        output: `/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

let someStatement;
/*
 ** some block comment not related to copyright.
 */
let someOtherComment`
      },
      {
        // Some comment
        code: `/*
 ** some comment in a file without copyright as first line
 */
let someStatement;`,
        errors: [
          {
            messageId: 'copyrightNeedsFix',
            type: 'Program'
          }
        ],
        output: `/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

/*
 ** some comment in a file without copyright as first line
 */
let someStatement;`
      },
      {
        // shebang and some comment
        code: `#!/usr/bin/env node
let someStatement;`,
        errors: [
          {
            messageId: 'copyrightNeedsFix',
            type: 'Program'
          }
        ],
        output: `#!/usr/bin/env node

/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

let someStatement;`
      },
      {
        // shebang and some comment
        code: `#!/usr/bin/env node
/*
 ** some block comment not related to copyright statement
 */
let someStatement;`,
        errors: [
          {
            messageId: 'copyrightNeedsFix',
            type: 'Program'
          }
        ],
        output: `#!/usr/bin/env node

/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

/*
 ** some block comment not related to copyright statement
 */
let someStatement;`
      },
      {
        // shebang and some comment
        code: `#!/usr/bin/env node
let someStatement;
/*
 ** some block comment not related to copyright statement
 */
let someOtherStatement;`,
        errors: [
          {
            messageId: 'copyrightNeedsFix',
            type: 'Program'
          }
        ],
        output: `#!/usr/bin/env node

/*
 ** Copyright (c) ${YEAR} Oracle and/or its affiliates.
 */

let someStatement;
/*
 ** some block comment not related to copyright statement
 */
let someOtherStatement;`
      }
    ]
  });
});
