// @oracle-cx-commerce/endpoints/collection/get-collection/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {createEndpoint, getBodyAsJson} from '@oracle-cx-commerce/endpoints/factory';
import {getAssetLanguageHeader, populateError, populateFieldParams} from '@oracle-cx-commerce/endpoints/utils';
import {getCategories} from '@oracle-cx-commerce/commerce-utils/selector';

export const processInput = ({categoryId, expandChildren, includeCollectionsWithNoActiveProducts, ...rest}) => {
  console.assert(categoryId, 'getCollection requires a category ID');

  return {
    headers: getAssetLanguageHeader(rest),
    params: [categoryId],
    query: {
      expand: expandChildren ? 'childCategories' : undefined,
      disableActiveProdCheck: includeCollectionsWithNoActiveProducts,
      ...rest,
      ...populateFieldParams(rest)
    }
  };
};

export const processOutput = async (response, state) => {
  const json = await getBodyAsJson(response);
  delete json.autoWrap;
  delete json.links;

  const categoryId = json.id;
  const collections = getCategories(state);
  delete collections[categoryId];

  return response.ok
    ? {
        catalogRepository: {
          categories: {
            [categoryId]: {
              ...json,
              route: json.route.replace(/^\//, '')
            }
          }
        }
      }
    : {
        catalogRepository: {categories: {...collections}},
        ...populateError(response, json)
      };
};

export default createEndpoint('getCollection', {
  processInput,
  processOutput
});
