// @oracle-cx-commerce/react-widgets/profile/account-address-book/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/profile/account-address-book/config';

const widgetResourceKeys = [
  'textAccountAddressBook',
  'textAccountAddress',
  'textProfileAddress',
  'labelAddNewAddress',
  'textNew',
  'textDefaultAddresses',
  'textProfileAddresses',
  'textInheritedAddresses',
  'textAddressesForThisAccount',
  'labelDefaultShippingAddress',
  'labelDefaultBillingAddress',
  'labelNoDefaultShippingAddressAvailable',
  'labelNoDefaultBillingAddressAvailable',
  'labelNoDefaultAddressesAvailable',
  'labelNoProfileAddressesAvailable',
  'labelNoAccountAddressesAvailable',
  'labelNoInheritedAddressesAvailable',
  'textAddressActions',
  'actionEdit',
  'actionDelete',
  'actionMakeDefaultShippingAddress',
  'actionRemoveDefaultShippingAddress',
  'actionMakeDefaultBillingAddress',
  'actionRemoveDefaultBillingAddress',
  'actionCopyAsProfileAddress',
  'actionCopyAsAccountAddress',
  'textLoading',
  'textLoadMoreAccountAddress',
  'textLoadMoreProfileAddress',
  'textLoadMoreInheritedAddress',
  'closeLinkAltText',
  'actionConfirm',
  'actionCancel',
  'headingDeleteProfileAddress',
  'headingDeleteAccountAddress',
  'textAddressDeletionMessage',
  'textConfirmAddressDeletionMessage',
  'searchIconAltText',
  'textFilterByNameOrAddress',
  'labelFirstName',
  'labelLastName',
  'labelCompanyName',
  'labelAddressLine1',
  'labelZipCode',
  'textDefaultAddressResultForSearchTerm',
  'textProfileAddressResultForSearchTerm',
  'textAccountAddressResultForSearchTerm',
  'textInheritedAddressResultForSearchTerm',
  'textDefaultAddressResultForSearchTermInSearchField',
  'textProfileAddressResultForSearchTermInSearchField',
  'textAccountAddressResultForSearchTermInSearchField',
  'textInheritedAddressResultForSearchTermInSearchField',
  'textAddressFilterFieldName'
];
export const AccountAddressBook = {
  packageId: '@oracle-cx-commerce/react-widgets',
  availableToAllPages: false,
  pageTypes: ['profile'],
  config: mergeDefaultConfig(config),
  resources: buildResources(resourceBundle, widgetResourceKeys),
  actions: ['listProfileAddresses', 'listOrganizationAddresses', 'getOrganizationDefaultAddresses']
};
