// @oracle-cx-commerce/actions/access-role/meta.js
/*
 ** Copyright (c) 2021 Oracle and/or its affiliates.
 */

export const createRole = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createRole'],
  description: 'Create Role. Creates a new role.',
  input: '@oracle-cx-commerce/actions/access-role/schema/create-role/input.json'
};

export const getRole = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getRole'],
  description: 'Get Role. Returns a role by ID.',
  input: '@oracle-cx-commerce/actions/access-role/schema/get-role/input.json'
};

export const updateRole = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateRole'],
  description: 'Update Role. Updates a given role by ID.',
  input: '@oracle-cx-commerce/actions/access-role/schema/update-role/input.json'
};

export const listRoles = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listRoles'],
  description: 'List Roles. Get all roles for shoppers.',
  input: '@oracle-cx-commerce/actions/access-role/schema/list-roles/input.json'
};

export const listAccessRights = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listAccessRights'],
  description: 'List Access Rights. Get all access rights for shoppers.',
  input: '@oracle-cx-commerce/actions/access-role/schema/list-access-rights/input.json'
};
