// @oracle-cx-commerce/react-widgets/cart/cart-pay-in-store/selectors.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import {
  isInStorePaymentApplicable,
  isPaymentEnabledForApproval,
  isPaymentTypeEnabled,
  isPayLaterSelected,
  isZeroValueOrder,
  getPaymentEnabledForScheduledOrder
} from '@oracle-cx-commerce/react-components/utils/payment';
import {getCurrentOrder, getCurrentOrderId, getCurrentProfile} from '@oracle-cx-commerce/commerce-utils/selector';
import {PAYMENT_TYPE_PAY_IN_STORE} from '@oracle-cx-commerce/commerce-utils/constants';

/**
 * Returns widget's required data
 * @param {Object} state the state object
 * @returns required data
 */
export const getComponentData = state => {
  // selectors
  const order = getCurrentOrder(state);
  //determines widget visibility
  const isDisplayCartPayInStore =
    isPaymentTypeEnabled(state, PAYMENT_TYPE_PAY_IN_STORE) &&
    isInStorePaymentApplicable(order) &&
    !isZeroValueOrder(order);

  return {
    isApprovalEnabled: getCurrentProfile(state).derivedApprovalRequired,
    isPaymentMethodEnabledForApproval: isPaymentEnabledForApproval(state, PAYMENT_TYPE_PAY_IN_STORE),
    currentOrderId: getCurrentOrderId(state),
    isPayAfterApprovalSelected: isPayLaterSelected(state),
    isDisplayCartPayInStore,
    PaymentInfoForScheduledOrder: getPaymentEnabledForScheduledOrder(state, PAYMENT_TYPE_PAY_IN_STORE)
  };
};
