// @oracle-cx-commerce/react-widgets/common/placeholder-stack/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import resourcesDE from './translations/de.json';
import resourcesEN from './translations/en.json';

const type = 'stack';

const config = {
  maxVariants: 10,
  minVariants: 1,
  minWidth: 1,
  canEditSubRegion: true,
  availableToAllPages: false,
  pageTypes: ['home']
};

const locales = {
  en: {
    resources: {
      displayName: resourcesEN.displayName,
      nextButton: resourcesEN.nextButton,
      previousButton: resourcesEN.previousButton
    }
  },
  de: {
    resources: {
      displayName: resourcesDE.displayName,
      nextButton: resourcesDE.nextButton,
      previousButton: resourcesDE.previousButton
    }
  },
  'en-GB': {
    resources: {
      displayName: resourcesEN.displayName,
      nextButton: resourcesEN.nextButton,
      previousButton: resourcesEN.previousButton
    }
  }
};

const regions = [
  {
    name: 'region1',
    locales: {
      en: {resources: {displayName: resourcesEN.step1}},
      de: {resources: {displayName: resourcesDE.step1}}
    }
  },
  {
    name: 'region2',
    locales: {
      en: {resources: {displayName: resourcesEN.step2}},
      de: {resources: {displayName: resourcesDE.step2}}
    }
  },
  {
    name: 'region3',
    locales: {
      en: {resources: {displayName: resourcesEN.step3}},
      de: {resources: {displayName: resourcesDE.step3}}
    }
  }
];

export const PlaceholderStack = {type, config, regions, locales};
