// @oracle-cx-commerce/endpoints/product/list-products/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {
  createCommaSeparatedStringFromArray,
  getAssetLanguageHeader,
  populateError,
  populateFieldParams
} from '@oracle-cx-commerce/endpoints/utils';
import {createEndpoint, getBodyAsJson} from '@oracle-cx-commerce/endpoints/factory';
import {transformProducts} from '@oracle-cx-commerce/endpoints/utils/transform';

export const processInput = ({productIds, search, priceListGroupId, ...rest}) => {
  return {
    headers: getAssetLanguageHeader(rest),
    query: {
      productIds: createCommaSeparatedStringFromArray(productIds),
      q: search,
      storePriceListGroupId: priceListGroupId,
      ...rest,
      ...populateFieldParams(rest)
    }
  };
};

export const processOutput = async (response, state) => {
  const json = await getBodyAsJson(response);
  delete json.autoWrap;
  delete json.links;

  // Add the retrieved products to the catalogRepository
  return response.ok ? transformProducts(json, state) : populateError(response, json);
};

export default createEndpoint('listProducts', {
  processInput,
  processOutput
});
