// @oracle-cx-commerce/react-widgets/product-listing/category-image-carousel/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const configResourceKeys = [
  'configTagTypeDropdownLabel',
  'configTagTypeDropdownHelpText',
  'configHeroTagLabel',
  'configSmallTagLabel',
  'configMediumTagLabel',
  'configLargeTagLabel',
  'configThumbnailTagLabel',
  'configIsAutoSlideLabel',
  'configAutoSlideImagesIntervalLabel',
  'configAutoSlideHelpText'
];

const config = mergeDefaultConfig({
  properties: [
    {
      id: 'tagTypeDropdown',
      type: 'optionType',
      labelResourceId: 'configTagTypeDropdownLabel',
      helpTextResourceId: 'configTagTypeDropdownHelpText',
      defaultValue: 'heroTag',
      required: true,
      options: [
        {
          id: 'heroTag',
          value: 'heroTag',
          labelResourceId: 'configHeroTagLabel'
        },
        {
          id: 'smallTag',
          value: 'smallTag',
          labelResourceId: 'configSmallTagLabel'
        },
        {
          id: 'mediumTag',
          value: 'mediumTag',
          labelResourceId: 'configMediumTagLabel'
        },
        {
          id: 'largeTag',
          value: 'largeTag',
          labelResourceId: 'configLargeTagLabel'
        },
        {
          id: 'thumbnailTag',
          value: 'thumbnailTag',
          labelResourceId: 'configThumbnailTagLabel'
        }
      ]
    },
    {
      id: 'isAutoSlide',
      type: 'booleanType',
      labelResourceId: 'configIsAutoSlideLabel',
      helpTextResourceId: 'configAutoSlideHelpText',
      defaultValue: true
    },
    {
      id: 'autoSlideImagesInterval',
      type: 'stringType',
      labelResourceId: 'configAutoSlideImagesIntervalLabel',
      helpTextResourceId: 'configAutoSlideHelpText',
      defaultValue: '3',
      pattern: '^[1-9]\\d*$'
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
});

export default config;
