// @oracle-cx-commerce/commerce-utils/selector/wishlists.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import {getSessionContext} from '@oracle-cx-commerce/commerce-utils/selector';

/**
 * @type {import('.').Selector}
 */
export const getWishlistRepository = state => state.wishlistRepository || {};

/**
 * @type {import('.').Selector}
 */
export const wishlists = state => getWishlistRepository(state).spaces || {};

export const getWishlists = state => getWishlistRepository(state).spaces || {};

export const getWishList = (state, {id}) => getWishlists(state)[id] || {};

// export const getWishListIds = state => Object.keys(getWishlists(state)) || {};

// export const getWishlistMembers = state => {
//   const {members} = wishlists(state);

export const getWishlistMembers = state => getWishlistRepository(state).users || {};

export const getWishlistMembersForSpaceId = (state, {id}) => getWishlistRepository(state).members[id] || {};

/**
 * @type {import('.').Selector}
 */
// export const isWishlistAccessTokenAvailable = state => ((access(state).access_token ? true : false));

//new selectors
export const getUsers = state => getWishlistRepository(state).users || {};

export const getLoggedInWishListUserId = state => getSessionContext(state).wishListUserId || '';

export const getLoggedInUser = (state, {wishListUserId = getLoggedInWishListUserId(state)} = {}) =>
  getUsers(state)[wishListUserId] || {};

export const getWishlistAccessToken = state => getLoggedInUser(state).access_token || '';

export const getWishlistSiteId = state => getLoggedInUser(state).siteId || '';

export const getSites = state => getWishlistRepository(state).sites || {};

export const getWishListsForCurrentSite = (state, {siteId = getWishlistSiteId(state)} = {}) =>
  getSites(state)[siteId] || {};

export const getWishListIds = state => getWishListsForCurrentSite(state).items || [];

export const isWishlistAccessTokenAvailable = state => (getLoggedInUser(state).access_token ? true : false);

/**
 * @type {import('.').Selector}
 */
export const access = state => getUsers(state)[getLoggedInWishListUserId(state)] || {};

export const wishListSiteId = state => access(state).sitedId || '';

export const getWishlistIdsForCurrentSite = state => getWishListsForCurrentSite(state).items || [];
