// @oracle-cx-commerce/react-widgets/profile/contact-registration-requests/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import contactRegistrationRequestConfig from '@oracle-cx-commerce/react-widgets/profile/contact-registration-requests/config';

export const widgetResourceKeys = [
  'actionShowMoreRegistrationRequests',
  'headingRegistrationRequests',
  'labelAriaFilterText',
  'labelAriaSortText',
  'labelRequestDate',
  'labelRequestId',
  'labelSite',
  'labelStatus',
  'searchIconAltText',
  'textAllStatuses',
  'textEmailAsc',
  'textEmailDesc',
  'textFilterByLastName',
  'textFirstNameAsc',
  'textFirstNameDesc',
  'textLastNameAsc',
  'textLastNameDesc',
  'textMoreInfoRequired',
  'textNameNotRetained',
  'textNoRegistrationRequest',
  'textRegistrationRequestRejected',
  'textRegistrationRequestsNew',
  'textRegistrationRequestsReview',
  'textRegistrationRequestSummary',
  'textRequestDateNewest',
  'textRequestDateOldest',
  'textRequestIDAsc',
  'textRequestIdDesc',
  'textRetrievingRegistrationRequests',
  'textStatusAsc',
  'textStatusDesc'
];

export const ContactRegistrationRequests = {
  packageId: '@oracle-cx-commerce/react-widgets',
  pageTypes: ['contact-registration-requests'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: contactRegistrationRequestConfig
};
