// @oracle-cx-commerce/react-widgets/checkout/checkout-gift-card/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';

const configResourceKeys = [
  'configCollapsedText',
  'configExpandedText',
  'configGiftCardDefaultStateHelpText',
  'configGiftCardDefaultStateLabel',
  'configIsPinRequiredHelpText',
  'configIsPinRequiredLabel'
];

const config = {
  properties: [
    {
      id: 'giftCardDefaultState',
      type: 'optionType',
      labelResourceId: 'configGiftCardDefaultStateLabel',
      helpTextResourceId: 'configGiftCardDefaultStateHelpText',
      defaultValue: 'collapsed',
      options: [
        {
          id: 'collapsed',
          value: 'collapsed',
          labelResourceId: 'configCollapsedText'
        },
        {
          id: 'expanded',
          value: 'expanded',
          labelResourceId: 'configExpandedText'
        }
      ],
      required: true
    },
    {
      id: 'isPinRequired',
      type: 'booleanType',
      labelResourceId: 'configIsPinRequiredLabel',
      helpTextResourceId: 'configIsPinRequiredHelpText',
      defaultValue: true,
      required: true
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
};

export default config;
