// @oracle-cx-commerce/react-widgets/checkout/checkout-gift-card/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/checkout/checkout-gift-card/config';

const widgetResourceKeys = [
  'actionApply',
  'actionRemove',
  'actionCheckBalance',
  'labelApplyAGiftCard',
  'labelGiftCardNumber',
  'labelGiftCardPin',
  'labelRemoveGiftCard',
  'textDuplicateGiftCard',
  'textNoBalanceGiftCard',
  'textRequiredField',
  'textGiftCardAdded',
  'textYourGiftCardBalance',
  'textUseGiftCardBalance',
  'actionConfirm',
  'headingConflictingPaymentMethods',
  'actionCancel',
  'alertForGiftCardPaymentWhenInstorePaymentApplied',
  'alertDoYouWantToApplyGiftCard',
  'closeLinkAltText',
  'textGiftCardPendingPaymentNote',
  'textGiftCardAddedToOrder',
  'alertOrderNotPlacedPaymentDeclined',
  'alertTechnicalProblemContactUs'
];

export const CheckoutGiftCard = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  availableToAllPages: false,
  pageTypes: ['checkout-payment', 'pending-payment'],
  config: mergeDefaultConfig(config),
  actions: ['inquireBalance', 'clearInquireBalance', 'applyPayments', 'addPayments', 'deleteAppliedPayment']
};
