// @oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/date-field/index.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import React, {useEffect, useState} from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/date-field/styles.css';
import PropTypes from 'prop-types';

const SYSTEM_CURRENT_DATE = '_system_current_date';
/**
 * This component will render a date UI control. It assumes that defaultValue
 * contains a date in ISO format i.e. YYYY-MM-DD
 *
 * @param {Object} props.inputRef
 *   A reference to the DOM element created in the render method. This will be
 *   passed from the parent component.
 * @param {String} props.uniqueId
 *   A unique id for the rendered DOM element, it should be unique within the HTML
 *   document. This will be passed from the parent component.
 * @param {String} props.name
 *   A display name (label) for the form control. This will be passed from the
 *   parent component.
 * @param {String} props.variableName
 *   The variable name used to connnect a form control to its corresponding value in
 *   configData . This will be passed from the parent component.
 * @param {String} props.defaultValue
 *   The initial value for the form control. It is retrieved from the configData
 *   object from the configurations repository store. This will be passed from
 *   the parent component. See:
 *   https://reactjs.org/docs/uncontrolled-components.html#default-values
 * @param {Function} props.handleChange
 *   The function to be called when a user has changed the value of the form control.
 *   This will be passed from the parent component.
 * @param {Boolean} props.updatable
 *   If false the form control should be disabled / read only. This will be
 *   passed from the parent component.
 * @param {Boolean} props.required
 *   If true the form control should be a mandatory / required field. This will be
 *   passed from the parent component.
 * @param {String} props.minimum
 *   The minmium value allowed to be input for the form control. This will be
 *   passed from the parent component.
 * @param {String} props.maximum
 *   The maximum value allowed to be input for the form control. This will be
 *   passed from the parent component.
 */
const DateField = ({
  inputRef,
  uniqueId,
  name,
  variableName,
  defaultValue,
  handleChange,
  updatable,
  required,
  minimum,
  maximum
}) => {
  const [minimumDate, setMinimumDate] = useState(null);

  useEffect(() => {
    if (minimum) {
      if (minimum === SYSTEM_CURRENT_DATE) {
        setMinimumDate(new Date().toISOString().slice(0, 10));
      } else {
        setMinimumDate(minimum);
      }
    } else {
      setMinimumDate(null);
    }

    if (inputRef && inputRef.current) {
      inputRef.current.value = defaultValue;
    }
  }, [defaultValue, inputRef, minimum]);

  /**
   *
   */
  const handleDateChange = changeEvent => {
    const newValue = changeEvent.target.value;
    if (newValue !== defaultValue) {
      handleChange(newValue);
    }
  };

  return (
    <Styled id="ConfiguratorFormControlDate" css={css}>
      <div className="ConfiguratorFormControlDate__Container">
        <label htmlFor={uniqueId}>{name}</label>
        <input
          id={uniqueId}
          type="date"
          name={variableName}
          min={minimumDate}
          max={maximum}
          onBlur={handleDateChange}
          defaultValue={defaultValue}
          ref={inputRef}
          required={required}
          disabled={!updatable}
        />
      </div>
    </Styled>
  );
};

DateField.propTypes = {
  /**
   * A reference to the DOM element created in the render method.
   * This will be passed from the parent component.
   */
  inputRef: PropTypes.shape(Object),
  /**
   * A unique id for the rendered DOM element, it should be unique within the HTML
   * document. This will be passed from the parent component.
   */
  uniqueId: PropTypes.string.isRequired,
  /**
   * A display name (label) for the form control. This will be passed from the
   * parent component.
   */
  name: PropTypes.string.isRequired,
  /**
   * The variable name used to connect a form control to its corresponding value in
   * configData . This will be passed from the parent component.
   */
  variableName: PropTypes.string.isRequired,
  /**
   * The initial value for the form control. It is retrieved from the configData
   * object from the configurations repository store. This will be passed from
   * the parent component. See:
   * https://reactjs.org/docs/uncontrolled-components.html#default-values
   */
  defaultValue: PropTypes.string.isRequired,
  /**
   * The function to be called when a user has changed the value of the form control.
   * This will be passed from the parent component.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * If false the form control should be disabled / read only. This will be
   * passed from the parent component.
   */
  updatable: PropTypes.bool.isRequired,
  /**
   * If true the form control should be a mandatory / required field. This will be
   * passed from the parent component.
   */
  required: PropTypes.bool.isRequired,
  /**
   * The minimum value allowed to be input for the form control. This will be
   * passed from the parent component.
   */
  minimum: PropTypes.string,
  /**
   * The maximum value allowed to be input for the form control. This will be
   * passed from the parent component.
   */
  maximum: PropTypes.string
};

DateField.defaultProps = {
  inputRef: null,
  maximum: undefined,
  minimum: undefined
};

export default DateField;
