// @oracle-cx-commerce/endpoints/product/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addProductNotification = () => import('@oracle-cx-commerce/endpoints/product/add-product-notification');
export const getProduct = () => import('@oracle-cx-commerce/endpoints/product/get-product');
export const getProductType = () => import('@oracle-cx-commerce/endpoints/product/get-product-type');
export const getProductsPrices = () => import('@oracle-cx-commerce/endpoints/product/get-products-prices');
export const getSkusPrices = () => import('@oracle-cx-commerce/endpoints/product/get-skus-prices');
export const getStockStatus = () => import('@oracle-cx-commerce/endpoints/product/get-stock-status');
export const listProducts = () => import('@oracle-cx-commerce/endpoints/product/list-products');
export const listProductTypes = () => import('@oracle-cx-commerce/endpoints/product/list-product-types');
export const listSkus = () => import('@oracle-cx-commerce/endpoints/product/list-skus');
export const moreProducts = () => import('@oracle-cx-commerce/endpoints/product/more-products');
