// @oracle-cx-commerce/endpoints/product/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export {default as addProductNotification} from '@oracle-cx-commerce/endpoints/product/add-product-notification/meta';
export {default as getProduct} from '@oracle-cx-commerce/endpoints/product/get-product/meta';
export {default as getProductType} from '@oracle-cx-commerce/endpoints/product/get-product-type/meta';
export {default as getProductsPrices} from '@oracle-cx-commerce/endpoints/product/get-products-prices/meta';
export {default as getSkusPrices} from '@oracle-cx-commerce/endpoints/product/get-skus-prices/meta';
export {default as getStockStatus} from '@oracle-cx-commerce/endpoints/product/get-stock-status/meta';
export {default as listProducts} from '@oracle-cx-commerce/endpoints/product/list-products/meta';
export {default as listProductTypes} from '@oracle-cx-commerce/endpoints/product/list-product-types/meta';
export {default as listSkus} from '@oracle-cx-commerce/endpoints/product/list-skus/meta';
export {default as moreProducts} from '@oracle-cx-commerce/endpoints/product/more-products/meta';
