// @oracle-cx-commerce/actions/merchant-settings/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const getCancelReasons = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCancelReasons'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-cancel-reasons/input.json',
  description:
    'Get cancel Reasons. Lists the cancel reasons that can be given while receiving items against a cancel request'
};

export const getSamlSettings = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getSamlSettings'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-saml-settings/input.json',
  description: 'Get Saml Settings. Get the SAML settings'
};

export const getProductionExperiments = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductionExperiments'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-production-experiments/input.json',
  description: 'fetches the production experiments data from server and stores it in merchant Repository'
};

export const getProductionFacebook = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductionFacebook'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-production-facebook/input.json',
  description: 'fetches the production facebook data from server and stores it in merchant Repository'
};

export const getProductionRecommendations = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductionRecommendations'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-production-recommendations/input.json'
};

export const getCloudConfiguration = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCloudConfiguration'],
  input: '@oracle-cx-commerce/actions/merchant-settings/schema/get-cloud-configuration/input.json',
  description: 'Get Cloud Configuration. Gets the configurations from server.'
};
