// @oracle-cx-commerce/actions/cart-gwp/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addGiftWithPurchaseItemToCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addGiftWithPurchaseItemToCart'],
  description: 'Add GiftWithPurchase Item to cart.',
  input: '@oracle-cx-commerce/actions/cart-gwp/schema/add-gift-with-purchase-item-to-cart/input.json'
};

export const changeGiftWithPurchaseItemInCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['changeGiftWithPurchaseItemInCart'],
  description: 'Change GiftWithPurchase Item In Cart.',
  input: '@oracle-cx-commerce/actions/cart-gwp/schema/change-gift-with-purchase-item-in-cart/input.json'
};

export const removeGiftWithPurchasePromotionFromCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['removeGiftWithPurchasePromotionFromCart'],
  description: 'Update Gift With Purchase Promotion On Cart',
  input: '@oracle-cx-commerce/actions/cart-gwp/schema/remove-gift-with-purchase-promotion-from-cart/input.json'
};

export const removeAllGiftWithPurchasePromotionsFromCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['removeAllGiftWithPurchasePromotionsFromCart'],
  description: 'Remove all Unselected Gifts',
  input: '@oracle-cx-commerce/actions/cart-gwp/schema/remove-all-gift-with-purchase-promotions-from-cart/input.json'
};

export const getGiftWithPurchaseItemChoicesForCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getGiftWithPurchaseItemChoicesForCart'],
  description: 'List of Selectable Gifts Available.',
  input: '@oracle-cx-commerce/actions/cart-gwp/schema/get-gift-with-purchase-item-choices-for-cart/input.json'
};
