// @oracle-cx-commerce/actions/cart-shipping-group/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addShippingGroupsToCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addShippingGroupsToCart'],
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/add-shipping-groups-to-cart/input.json',
  description: 'Add Shipping Groups. API to add Shipping groups to the current incomplete order.'
};

export const deleteShippingGroupFromCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteShippingGroupFromCart'],
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/delete-shipping-group-from-cart/input.json',
  description:
    "Delete Shipping Group. Delete Shipping group from an order. Shipping group can be deleted from the order only if no relationship exists for that shippinggroup. Optionally takes the x-ccasset-language headers to get translated content in another language. It has an optional X-CCVisitorId header param which will be used in case of anonumous user. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.This endpoint returns the order details irrespective of organization ID passed in request header. If user doesn't belong to order's organization, then the user has no access to delete items from order."
};

export const deleteShippingGroupsFromCart = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteShippingGroupsFromCart'],
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/delete-shipping-groups-from-cart/input.json',
  description: 'Delete multiple shipping groups of the current order'
};

export const getCartShippingGroup = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCartShippingGroup'],
  description: 'Get given single shipping group of the current order',
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/get-cart-shipping-group/input.json'
};

export const getCartShippingGroups = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCartShippingGroups'],
  description: 'Get list of shipping groups of the current order',
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/get-cart-shipping-groups/input.json'
};

export const relateShippingGroupsToCartItems = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['relateShippingGroupsToCartItems'],
  description: 'API used to relate shipping groups to commerce items. Schema is incomplete',
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/relate-shipping-groups-to-cart-items/input.json'
};

export const updateCartShippingGroup = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateCartShippingGroup'],
  description: 'Update Shipping Group. Update single Shipping group of the current incomplete order.',
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/update-cart-shipping-group/input.json'
};

export const updateCartShippingGroups = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateCartShippingGroups'],
  description: 'Update Shipping Groups. Update Shipping groups of the current incomplete order.',
  input: '@oracle-cx-commerce/actions/cart-shipping-group/schema/update-cart-shipping-groups/input.json'
};
