// @oracle-cx-commerce/react-widgets/product/product-recommendations-carousel/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const configResourceKeys = [
  'configMaximumNumberOfProductsLabel',
  'configMaximumNumberOfProductsHelpText',
  'configRecommendationsStrategyLabel',
  'configRecommendationsStrategyHelpText'
];

const config = mergeDefaultConfig({
  properties: [
    {
      id: 'maximumNumberOfProducts',
      type: 'stringType',
      labelResourceId: 'configMaximumNumberOfProductsLabel',
      defaultValue: '12',
      helpTextResourceId: `configMaximumNumberOfProductsHelpText`,
      required: true,
      pattern: '^[1-9]\\d*$'
    },
    {
      id: 'strategy',
      type: 'select2Type',
      name: 'strategy',
      helpTextResourceId: 'configRecommendationsStrategyHelpText',
      labelResourceId: 'configRecommendationsStrategyLabel',
      required: true,
      dataKey: 'strategies',
      defaultValue: 'Blended'
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys),
  defaults: {
    maximumNumberOfProducts: '12',
    strategy: 'Blended'
  }
});

export default config;
