// @oracle-cx-commerce/react-widgets/profile/profile-edit-delete-scheduled-order/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import {widgetResourceKeys as checkoutScheduledOrderKeys} from '@oracle-cx-commerce/react-widgets/checkout/checkout-scheduled-order/meta';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'textDeleteScheduleWarning',
  'textDeleteScheduledOrder',
  'textEditSchedule',
  'actionCancel',
  'actionSave',
  'actionConfirm',
  'closeLinkAltText',
  'actionShowMore',
  'textDeleteScheduledOrderConfirm',
  'textAllFieldsRequired',
  ...checkoutScheduledOrderKeys
];

export const ProfileEditDeleteScheduledOrder = {
  packageId: '@oracle-cx-commerce/react-widgets',
  availableToAllPages: false,
  pageTypes: ['Scheduled Order Details'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  requiresContext: ['container_context'],
  config: defaultConfig
};
