// @oracle-cx-commerce/react-widgets/checkout/checkout-multi-shipping-with-bopis-details/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/checkout/checkout-multi-shipping-with-bopis-details/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import {widgetResourceKeys as CheckoutMultiShippingDetailsResourceKey} from '@oracle-cx-commerce/react-widgets/checkout/checkout-multi-shipping-details/meta';
import {widgetResourceKeys as ExtendedCartItemDetailsResourceKey} from '@oracle-cx-commerce/react-widgets/cart/extended-cart-item-details/meta';

const widgetResourceKeys = [
  ...CheckoutMultiShippingDetailsResourceKey,
  ...ExtendedCartItemDetailsResourceKey,
  'labelHomeDeliveryItems',
  'labelHomeDeliveryAddress',
  'labelInStorePickUp',
  'labelChange',
  'actionShipHomeDeliveryItemsToMultipleAddresses',
  'labelHomeDeliveryAddresses',
  'actionShipHomeDeliveryItemsToASingleAddress'
];

export const CheckoutMultiShippingWithBOPISDetails = {
  availableToAllPages: false,
  config: mergeDefaultConfig(config),
  packageId: '@oracle-cx-commerce/react-widgets',
  pageTypes: ['checkout-shipping'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  requiresContext: ['checkout_shipping_context', 'cart_context']
};
