// @oracle-cx-commerce/actions/product/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const addProductNotification = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['addProductNotification'],
  description:
    'Create a new product notification, which sends an email to a user when the specified product comes back in stock.',
  input: '@oracle-cx-commerce/actions/product/schema/add-product-notification/input.json'
};

export const getProduct = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProduct'],
  description: 'Get a product details.',
  input: '@oracle-cx-commerce/actions/product/schema/get-product/input.json'
};

export const getProductType = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductType'],
  description: 'Get Product Type. Returns all the product types.',
  input: '@oracle-cx-commerce/actions/product/schema/get-product-type/input.json'
};

export const getProductsPrices = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductsPrices'],
  description:
    'Get Products Prices. Return the the list, sale and complex prices for the list of products. Optionally takes the -ccasset-language header to get translated content in another language.\n',
  input: '@oracle-cx-commerce/actions/product/schema/get-products-prices/input.json'
};

export const getSkusPrices = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getSkusPrices'],
  description: 'Get SKUs Prices. Return the the list, sale and complex prices for the list of SKUs.',
  input: '@oracle-cx-commerce/actions/product/schema/get-skus-prices/input.json'
};

export const getStockStatus = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getStockStatus'],
  description: 'Get the stock status for an array of SKUs or products.',
  input: '@oracle-cx-commerce/actions/product/schema/get-stock-status/input.json'
};

export const listProductTypes = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listProductTypes'],
  description: 'List Product Types. Returns all the product types.',
  input: '@oracle-cx-commerce/actions/product/schema/list-product-types/input.json'
};

export const listSkus = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listSkus'],
  description: 'Gets the data and related selected options of a list of SKUs.',
  input: '@oracle-cx-commerce/actions/product/schema/list-skus/input.json'
};

export const listProducts = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listProducts'],
  description: 'List Products. Get a list of products. None of the image URLS are encoded.',
  input: '@oracle-cx-commerce/actions/product/schema/list-products/input.json'
};

export const moreProducts = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['moreProducts'],
  description:
    'Get More Products. Get more products specifically to render on the current page.  The category is detected from the current page route',
  input: '@oracle-cx-commerce/actions/product/schema/more-products/input.json'
};
