// @oracle-cx-commerce/commerce-utils/constants/address.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
export const VALIDATION_CONFIG = {
  addressType: {
    name: 'addressType',
    maxLength: 50,
    required: true,
    disabled: false
  },
  firstName: {
    name: 'firstName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  lastName: {
    name: 'lastName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  companyName: {
    name: 'companyName',
    maxLength: 50,
    required: true,
    disabled: false
  },
  country: {
    name: 'country',
    required: true,
    disabled: false
  },
  postalCode: {
    name: 'postalCode',
    maxLength: 10,
    required: true,
    disabled: false
  },
  state: {
    name: 'state',
    required: true,
    disabled: false
  },
  address1: {
    name: 'address1',
    maxLength: 60,
    required: true,
    disabled: false
  },
  city: {
    name: 'city',
    maxLength: 50,
    required: true,
    disabled: false
  },
  phoneNumber: {
    name: 'phoneNumber',
    maxLength: 15,
    disabled: false,
    required: false,
    pattern: '[0-9()+ -]+'
  },
  isDefaultShippingAddress: {
    name: 'isDefaultShippingAddress',
    disabled: false
  },
  isDefaultBillingAddress: {
    name: 'isDefaultBillingAddress',
    disabled: false
  }
};

export const VALIDATION_CONFIG_BUYER = {
  addressType: {
    name: 'addressType',
    maxLength: 50,
    required: true,
    disabled: true
  },
  firstName: {
    name: 'firstName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  lastName: {
    name: 'lastName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  companyName: {
    name: 'companyName',
    maxLength: 50,
    required: true,
    disabled: true
  },
  country: {
    name: 'country',
    required: true,
    disabled: true
  },
  postalCode: {
    name: 'postalCode',
    maxLength: 10,
    required: true,
    disabled: true
  },
  state: {
    name: 'state',
    required: true,
    disabled: true
  },
  address1: {
    name: 'address1',
    maxLength: 60,
    required: true,
    disabled: true
  },
  city: {
    name: 'city',
    maxLength: 50,
    required: true,
    disabled: true
  },
  phoneNumber: {
    name: 'phoneNumber',
    maxLength: 15,
    disabled: true,
    required: false,
    pattern: '[0-9()+ -]+'
  },
  isDefaultShippingAddress: {
    name: 'isDefaultShippingAddress',
    disabled: true
  },
  isDefaultBillingAddress: {
    name: 'isDefaultBillingAddress',
    disabled: true
  }
};

export const VALIDATION_CONFIG_PROFILEADDRESSMANAGER = {
  addressType: {
    name: 'addressType',
    maxLength: 50,
    required: false,
    disabled: false
  },
  firstName: {
    name: 'firstName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  lastName: {
    name: 'lastName',
    maxLength: 60,
    required: true,
    disabled: false
  },
  companyName: {
    name: 'companyName',
    maxLength: 50,
    required: false,
    disabled: false
  },
  country: {
    name: 'country',
    disabled: false
  },
  postalCode: {
    name: 'postalCode',
    maxLength: 10,
    required: true,
    disabled: false
  },
  state: {
    name: 'state',
    required: true,
    disabled: false
  },
  address1: {
    name: 'address1',
    maxLength: 60,
    required: true,
    disabled: false
  },
  city: {
    name: 'city',
    maxLength: 50,
    required: true,
    disabled: false
  },
  phoneNumber: {
    name: 'phoneNumber',
    maxLength: 15,
    disabled: false,
    required: false,
    pattern: '[0-9()+ -]+'
  },
  isDefaultShippingAddress: {
    name: 'isDefaultShippingAddress',
    disabled: false
  },
  isDefaultBillingAddress: {
    name: 'isDefaultBillingAddress',
    disabled: false
  }
};
