// @oracle-cx-commerce/commerce-utils/constants/page-links.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const PAGE_401_LINK = '401';
export const PAGE_404_LINK = '404';
export const PAGE_500_LINK = '500';
export const PAGE_ACCOUNT_ADDRESS_BOOK_LINK = 'account-address-book';
export const PAGE_ACCOUNT_CONTACTS_LINK = 'account-contacts';
export const PAGE_ACCOUNT_CONTACT_HOME_LINK = 'home';
export const PAGE_ACCOUNT_HOME_LINK = 'home';
export const PAGE_ACCOUNT_PROFILE_LINK = 'profile';
export const PAGE_ADD_ACCOUNT_ADDRESS_LINK = 'add-account-address';
export const PAGE_ADD_ACCOUNT_CONTACT_LINK = 'add-account-contact';
export const PAGE_ADD_PROFILE_ADDRESS_LINK = 'add-profile-address';
export const PAGE_CANCEL_ORDER_LINK = 'cancel-order';
export const PAGE_CART_LINK = 'cart';
export const PAGE_CATEGORY_LINK = 'category';
export const PAGE_CHECKOUT_LOGIN_LINK = 'checkout-login';
export const PAGE_CHECKOUT_ORDER_CONFIRMATION_LINK = 'checkout-order-confirmation';
export const PAGE_CHECKOUT_PAYER_AUTHENTICATION_LINK = 'checkout-payer-authentication';
export const PAGE_CHECKOUT_PAYMENT_LINK = 'checkout-payment';
export const PAGE_CHECKOUT_REVIEW_ORDER_LINK = 'checkout-review-order';
export const PAGE_CHECKOUT_QUOTE_SHIPPING_LINK = 'checkout-quote-shipping';
export const PAGE_CHECKOUT_PAYMENT_AND_REVIEW_ORDER_LINK = 'checkout-quote-payment-and-review';
export const PAGE_CHECKOUT_SHIPPING_LINK = 'checkout-shipping';
export const PAGE_CHECKOUT_LINK = 'checkout';
export const PAGE_CONTACT_REQUEST_LINK = 'contact-request';
export const PAGE_CONTACT_REGISTRATION_REQUESTS_LINK = 'contact-registration-requests';
export const PAGE_CONTACT_US_LINK = 'contactus';
export const PAGE_CREATE_PURCHASE_LIST_LINK = 'create-purchase-list';
export const PAGE_CUSTOMER_SERVICES_LINK = 'customer-services';
export const PAGE_EDIT_ACCOUNT_ADDRESS_LINK = 'edit-account-address';
export const PAGE_EDIT_ACCOUNT_CONTACT_LINK = 'edit-account-contact';
export const PAGE_EDIT_PROFILE_ADDRESS_LINK = 'edit-profile-address';
export const PAGE_EMAIL_MARKETING_PREFERENCES_LINK = 'email-marketing-preferences';
export const PAGE_HELP_LINK = 'help';
export const PAGE_HOME_LINK = 'home';
export const PAGE_LOGIN_LINK = 'login';
export const PAGE_ORDER_APPROVAL_SETTINGS_LINK = 'order-approval-settings';
export const PAGE_ORDER_DETAILS_LINK = 'order-details';
export const PAGE_ORDER_HISTORY_LINK = 'order-history';
export const PAGE_ORDER_PENDING_APPROVAL_LINK = 'orders-pending-approval';
export const PAGE_PENDING_PAYMENT_LINK = 'pending-payment';
export const PAGE_PRODUCT_LINK = 'product';
export const PAGE_PROFILE_ADD_CREDIT_CARD_LINK = 'profile-add-credit-card';
export const PAGE_ADDRESS_BOOK_LINK = 'address-book';
export const PAGE_CREATE_RETURN_LINK = 'create-return';
export const PAGE_CONFIRM_DELETE_SAVED_CARD_LINK = 'confirm-delete-saved-card';
export const PAGE_PROFILE_DETAILS_LINK = 'profile-details';
export const PAGE_PROFILE_SAVED_CARDS_LINK = 'profile-saved-cards';
export const PAGE_PROFILE_LINK = 'profile';
export const PAGE_PURCHASE_LIST_DETAILS_LINK = 'purchase-list-details';
export const PAGE_PURCHASE_LISTS_LINK = 'purchase-lists';
export const PAGE_QUICK_ORDER_LINK = 'quick-order';
export const PAGE_REGISTRATION_LINK = 'registration';
export const PAGE_RESET_PASSWORD_LINK = 'reset-password';
export const PAGE_RETURN_DETAILS_LINK = 'return-details';
export const PAGE_RETURNS_REFUNDS_LINK = 'returns-refunds';
export const PAGE_SCHEDULED_ORDER_DETAILS_LINK = 'scheduled-order-details';
export const PAGE_SCHEDULED_ORDERS_LINK = 'scheduled-orders';
export const PAGE_SEARCH_LINK = 'search';
export const PAGE_SHIPPING_INFORMATION_LINK = 'shipping-information';
export const PAGE_UPDATE_EXPIRED_PASSWORD_LINK = 'update-expired-password';
export const PAGE_UPDATE_PASSWORD_LINK = 'update-password';
export const PAGE_WEB_CHECKOUT_RESPONSE_LINK = 'web-checkout-response';
export const PAGE_WISH_LISTS_LINK = 'wishlists';
export const PAGE_WISH_LIST_DETAILS_LINK = 'wish-list-details';

export const PAGINATION_PARAM_OFFSET = 'No';
export const PAGINATION_PARAM_LIMIT = 'Nrpp';
export const PAGINATION_PARAM_PAGE = 'page';

// pageTypes which should have canonical tags
export const CANONICAL_PAGE_TYPES = ['home', 'category', 'collection', 'product', 'article'];
// pageTypes which should have tags for next and previous products link
export const PRODUCT_LISTING_PAGE_TYPES = ['category', 'collection'];
