// @oracle-cx-commerce/commerce-utils/constants/payments.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
export const CYBERSOURCE_SOP = 'sop';
export const GIFT_CARD_NUMBER_PROP_NAME = 'giftCardNumber';
export const GIFT_CARD_PIN_PROP_NAME = 'giftCardPin';
export const GIFT_CARDS = 'giftCards';
export const PAYER_AUTH_REQUIRED = 'payerAuthRequired';
export const PAYMENT_AUTH_DECLINED = 'paymentAuthDeclined';
export const PAYMENT_AUTH_SUCCESS = 'paymentAuthSuccess';
export const PAYMENT_AUTH_TIMED_OUT = 'paymentAuthTimeout';
export const PAYMENT_AUTHORIZATION_DEFAULT_TRIES = 50;
export const PAYMENT_AUTHORIZATION_DEFAULT_INTERVAL = 3000;
export const PAYMENT_METHOD_CREDIT_CARD = 'creditCard';
export const PAYMENT_METHOD_CUSTOM_CURRENCY = 'customCurrencyPaymentGroup';
export const PAYMENT_METHOD_INVOICE_REQUEST = 'invoiceRequest';
export const PAYMENT_METHOD_ONLINE_PAYMENT_GROUP = 'onlinePaymentGroup';
export const PAYMENT_METHOD_TOKENIZED_CREDIT_CARD = 'tokenizedCreditCard';
export const PAYMENT_REVERSAL_FAILED = 'paymentReversalFailed';
export const PAYMENT_STATE_AUTHORIZED = 'AUTHORIZED';
export const PAYMENT_STATE_AUTHORIZE_FAILED = 'AUTHORIZE_FAILED';
export const PAYMENT_STATE_INITIAL = 'INITIAL';
export const PAYMENT_STATE_PAYMENT_DEFERRED = 'PAYMENT_DEFERRED';
export const PAYMENT_STATE_PENDING_AUTHORIZATION = 'PENDING_AUTHORIZATION';
export const PAYMENT_STATE_PAYMENT_REQUEST_ACCEPTED = 'PAYMENT_REQUEST_ACCEPTED';
export const PAYMENT_STATE_REMOVED = 'REMOVED';
export const PAYMENT_STATE_SETTLED = 'SETTLED';
export const PAYMENT_TYPE_CARD = 'card';
export const PAYMENT_TYPE_CASH = 'cash';
export const PAYMENT_TYPE_GENERIC = 'generic';
export const PAYMENT_TYPE_GIFTCARD = 'physicalGiftCard';
export const PAYMENT_TYPE_INVOICE = 'invoice';
export const PAYMENT_TYPE_LOYALTYPOINTS = 'loyaltyPoints';
export const PAYMENT_TYPE_PAY_IN_STORE = 'inStorePayment';
export const PAYMENT_TYPE_STORECREDIT = 'storeCredit';
export const PAYMENT_TYPE_PAY_LATER = 'payLater';
