// @oracle-cx-commerce/endpoints/collection/list-collections/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {
  createCommaSeparatedStringFromArray,
  populateError,
  populateFieldParams
} from '@oracle-cx-commerce/endpoints/utils';
import {createEndpoint, getBodyAsJson} from '@oracle-cx-commerce/endpoints/factory';

export const processInput = ({categoryIds, ...rest}) => {
  console.assert(categoryIds, 'listCollections requires a category ID list');

  return {
    query: {
      categoryIds: createCommaSeparatedStringFromArray(categoryIds),
      ...rest,
      ...populateFieldParams(rest)
    }
  };
};

export const processOutput = async response => {
  const json = await getBodyAsJson(response);

  return response.ok
    ? {
        catalogRepository: {
          categories: json.items.reduce((categoryList, category) => {
            categoryList[category.id] = {...category};

            return categoryList;
          }, {})
        }
      }
    : {...populateError(response, json)};
};

export default createEndpoint('listCollections', {
  processInput,
  processOutput
});
