// @oracle-cx-commerce/react-widgets/profile/profile-purchase-lists/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-purchase-lists/config';
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

const widgetResourceKeys = [
  'textPurchaseListItems',
  'headingPurchaseLists',
  'labelNewList',
  'headingMyLists',
  'headingSharedLists',
  'labelMyPurchaseListsFilter',
  'labelSharedPurchaseListsFilter',
  'labelLoadMore',
  'textNoMatchingPurchaseLists',
  'textPurchaseListsFilterResults',
  'textRetrievingPurchaseLists',
  'nameasc',
  'namedesc',
  'labelAriaSortText',
  'labelAriaFilterText',
  'textAllSharedLists',
  'textSharedWithEmail',
  'textSharedWithAccount',
  'textPurchaseListOwnerDescription',
  'textOwnerLastNameAsc',
  'textOwnerLastNameDesc',
  'textOwnerEmailAsc',
  'textOwnerEmailDesc',
  'textPrivate',
  'textShared'
];

export const ProfilePurchaseLists = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  actions: ['listPurchaseLists'],
  pageTypes: ['Purchase Lists']
};
