// @oracle-cx-commerce/react-widgets/profile/profile-order-summary/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/profile/profile-order-summary/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'headingOrderSummary',
  'textExcludingTax',
  'textFree',
  'textGiftCardBalance',
  'textIncludingTax',
  'textOrderDiscounts',
  'textShipping',
  'textShippingDiscount',
  'textShippingSurcharge',
  'textSubtotal',
  'textTax',
  'textTotal',
  'textPriceChangeMessage',
  'textConditionalTotal',
  'textRemainingTotal',
  'labelPayInStore',
  'labelInvoice',
  'textOrderSummaryGiftCard'
];

export const ProfileOrderSummary = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  pageTypes: ['profile'],
  requiresContext: ['order_context']
};
