// @oracle-cx-commerce/endpoints/profile/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createProfile = () => import('@oracle-cx-commerce/endpoints/profile/create-profile');
export const createProfileAddress = () => import('@oracle-cx-commerce/endpoints/profile/create-profile-address');
export const deleteProfileAddress = () => import('@oracle-cx-commerce/endpoints/profile/delete-profile-address');
export const getCurrentProfile = () => import('@oracle-cx-commerce/endpoints/profile/get-current-profile');
export const getProfileAddress = () => import('@oracle-cx-commerce/endpoints/profile/get-profile-address');
export const getProfileSavedCard = () => import('@oracle-cx-commerce/endpoints/profile/get-profile-saved-card');
export const getShopperSettings = () => import('@oracle-cx-commerce/endpoints/profile/get-shopper-settings');
export const listProfileAddresses = () => import('@oracle-cx-commerce/endpoints/profile/list-profile-addresses');
export const listProfileSavedCards = () => import('@oracle-cx-commerce/endpoints/profile/list-profile-saved-cards');
export const listProfileSavedCardsForCurrentSite = () =>
  import('@oracle-cx-commerce/endpoints/profile/list-profile-saved-cards-for-current-site');
export const listSiteProperties = () => import('@oracle-cx-commerce/endpoints/profile/list-site-properties');
export const removeSavedCardFromProfile = () =>
  import('@oracle-cx-commerce/endpoints/profile/remove-saved-card-from-profile');
export const resetPassword = () => import('@oracle-cx-commerce/endpoints/profile/reset-password');
export const saveCardToProfile = () => import('@oracle-cx-commerce/endpoints/profile/save-card-to-profile');
export const updateDefaultProfileAddress = () =>
  import('@oracle-cx-commerce/endpoints/profile/update-default-profile-address');
export const updateExpiredPassword = () => import('@oracle-cx-commerce/endpoints/profile/update-expired-password');
export const updateProfile = () => import('@oracle-cx-commerce/endpoints/profile/update-profile');
export const updateProfileAddress = () => import('@oracle-cx-commerce/endpoints/profile/update-profile-address');
export const updateProfileSavedCard = () => import('@oracle-cx-commerce/endpoints/profile/update-profile-saved-card');
export const updateSiteProperties = () => import('@oracle-cx-commerce/endpoints/profile/update-site-properties');
