// @oracle-cx-commerce/endpoints/profile/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export {default as createProfile} from '@oracle-cx-commerce/endpoints/profile/create-profile/meta';
export {default as createProfileAddress} from '@oracle-cx-commerce/endpoints/profile/create-profile-address/meta';
export {default as deleteProfileAddress} from '@oracle-cx-commerce/endpoints/profile/delete-profile-address/meta';
export {default as getCurrentProfile} from '@oracle-cx-commerce/endpoints/profile/get-current-profile/meta';
export {default as getProfileAddress} from '@oracle-cx-commerce/endpoints/profile/get-profile-address/meta';
export {default as getProfileSavedCard} from '@oracle-cx-commerce/endpoints/profile/get-profile-saved-card/meta';
export {default as getShopperSettings} from '@oracle-cx-commerce/endpoints/profile/get-shopper-settings/meta';
export {default as listProfileAddresses} from '@oracle-cx-commerce/endpoints/profile/list-profile-addresses/meta';
export {default as listProfileSavedCards} from '@oracle-cx-commerce/endpoints/profile/list-profile-saved-cards/meta';
export {default as listProfileSavedCardsForCurrentSite} from '@oracle-cx-commerce/endpoints/profile/list-profile-saved-cards-for-current-site/meta';
export {default as listSiteProperties} from '@oracle-cx-commerce/endpoints/profile/list-site-properties/meta';
export {default as removeSavedCardFromProfile} from '@oracle-cx-commerce/endpoints/profile/remove-saved-card-from-profile/meta';
export {default as resetPassword} from '@oracle-cx-commerce/endpoints/profile/reset-password/meta';
export {default as saveCardToProfile} from '@oracle-cx-commerce/endpoints/profile/save-card-to-profile/meta';
export {default as updateDefaultProfileAddress} from '@oracle-cx-commerce/endpoints/profile/update-default-profile-address/meta';
export {default as updateExpiredPassword} from '@oracle-cx-commerce/endpoints/profile/update-expired-password/meta';
export {default as updateProfile} from '@oracle-cx-commerce/endpoints/profile/update-profile/meta';
export {default as updateProfileAddress} from '@oracle-cx-commerce/endpoints/profile/update-profile-address/meta';
export {default as updateProfileSavedCard} from '@oracle-cx-commerce/endpoints/profile/update-profile-saved-card/meta';
export {default as updateSiteProperties} from '@oracle-cx-commerce/endpoints/profile/update-site-properties/meta';
