// @oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/single-select-menu/single-select-drop-down.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import React, {useEffect, useContext} from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import {ConfiguratorContext} from '@oracle-cx-commerce/react-widgets/product/product-configure/contexts';
import css from '@oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/single-select-menu/styles.css';
import PropTypes from 'prop-types';

/**
 * This component will render a single select menu grid UI control.
 *
 * @param {Object} props.inputRef
 *   A reference to the DOM element created in the render method. This will be
 *   passed from the parent component.
 * @param {String} props.uniqueId
 *   A unique id for the rendered DOM element, it should be unique within the HTML
 *   document. This will be passed from the parent component.
 * @param {String} props.name
 *   A display name (label) for the form control. This will be passed from the
 *   parent component.
 * @param {String} props.variableName
 *   The variable name used to connnect a form control to its corresponding value in
 *   configData . This will be passed from the parent component.
 * @param {String} props.defaultValue
 *   The initial value for the form control. It is retrieved from the configData
 *   object from the configurations repository store. This will be passed from
 *   the parent component. See:
 *   https://reactjs.org/docs/uncontrolled-components.html#default-values
 * @param {Function} props.handleChange
 *   The function to be called when a user has changed the value of the form control.
 *   This will be passed from the parent component.
 * @param {Boolean} props.updatable
 *   If false the form control should be disabled / read only. This will be
 *   passed from the parent component.
 * @param {Boolean} props.required
 *   If true the form control should be a mandatory / required field. This will be
 *   passed from the parent component.
 * @param {Array} props.menuItems
 *   The collection of menu items to be displayed by the form control. This will be
 *   passed from the parent component
 */
const SingleSelectDropDown = ({
  inputRef,
  uniqueId,
  name,
  variableName,
  defaultValue,
  handleChange,
  updatable,
  required,
  menuItems
}) => {
  const {
    resources: {textPleaseSelect = ''}
  } = useContext(ConfiguratorContext);

  // this is required for interact
  useEffect(() => {
    if (inputRef && inputRef.current) {
      for (let i = 0; i < inputRef.current.options.length; i++) {
        if (inputRef.current.options[i].value === defaultValue) {
          inputRef.current.options[i].selected = true;
          break;
        }
      }
    }
  }, [defaultValue, inputRef]);

  /**
   *
   */
  const handleSingleSelectMenuChange = changeEvent => {
    const newValue = {
      value: changeEvent.target[changeEvent.target.selectedIndex].value,
      displayValue: changeEvent.target[changeEvent.target.selectedIndex].textContent
    };

    if (newValue.value !== defaultValue) {
      handleChange(newValue);
    }
  };

  return (
    <Styled id="ConfiguratorSingleSelect" css={css}>
      <div className="ConfiguratorSingleSelect__Container">
        <label htmlFor={uniqueId}>{name}</label>
        <select
          id={uniqueId}
          name={variableName}
          onChange={handleSingleSelectMenuChange}
          defaultValue={defaultValue}
          required={required}
          ref={inputRef}
          disabled={!updatable}
        >
          <option label={textPleaseSelect}></option>
          {menuItems.map(item => {
            return (
              <option key={item.value} value={item.value}>
                {item.displayValue}
              </option>
            );
          })}
        </select>
      </div>
    </Styled>
  );
};

SingleSelectDropDown.propTypes = {
  /**
   * A reference to the DOM element created in the render method.
   * This will be passed from the parent component.
   */
  inputRef: PropTypes.shape(Object),
  /**
   * A unique id for the rendered DOM element, it should be unique within the HTML
   * document. This will be passed from the parent component.
   */
  uniqueId: PropTypes.string.isRequired,
  /**
   * A display name (label) for the form control. This will be passed from the
   * parent component.
   */
  name: PropTypes.string.isRequired,
  /**
   * The variable name used to connect a form control to its corresponding value in
   * configData . This will be passed from the parent component.
   */
  variableName: PropTypes.string.isRequired,
  /**
   * The initial value for the form control. It is retrieved from the configData
   * object from the configurations repository store. This will be passed from
   * the parent component. See:
   * https://reactjs.org/docs/uncontrolled-components.html#default-values
   */
  defaultValue: PropTypes.string,
  /**
   * The function to be called when a user has changed the value of the form control.
   * This will be passed from the parent component.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * If false the form control should be disabled / read only. This will be
   * passed from the parent component.
   */
  updatable: PropTypes.bool.isRequired,
  /**
   * If true the form control should be a mandatory / required field. This will be
   * passed from the parent component.
   */
  required: PropTypes.bool.isRequired,
  /**
   * The collection of menu items to be displayed by the form control. This will be
   * passed from the parent component
   */
  menuItems: PropTypes.arrayOf(Object).isRequired
};

SingleSelectDropDown.defaultProps = {
  inputRef: null,
  defaultValue: null
};

export default SingleSelectDropDown;
