// @oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/single-select-menu/single-select-radio-button-horizontal.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import React, {useEffect} from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/product/product-configure/components/form-control/single-select-menu/styles.css';
import PropTypes from 'prop-types';
import Radio from '@oracle-cx-commerce/react-components/radio';

/**
 * This component will render a single select menu grid UI control.
 *
 * @param {Object} props.inputRef
 *   A reference to the DOM element created in the render method. This will be
 *   passed from the parent component.
 * @param {String} props.uniqueId
 *   A unique id for the rendered DOM element, it should be unique within the HTML
 *   document. This will be passed from the parent component.
 * @param {String} props.name
 *   A display name (label) for the form control. This will be passed from the
 *   parent component.
 * @param {String} props.variableName
 *   The variable name used to connnect a form control to its corresponding value in
 *   configData . This will be passed from the parent component.
 * @param {String} props.defaultValue
 *   The initial value for the form control. It is retrieved from the configData
 *   object from the configurations repository store. This will be passed from
 *   the parent component. See:
 *   https://reactjs.org/docs/uncontrolled-components.html#default-values
 * @param {Function} props.handleChange
 *   The function to be called when a user has changed the value of the form control.
 *   This will be passed from the parent component.
 * @param {Boolean} props.updatable
 *   If false the form control should be disabled / read only. This will be
 *   passed from the parent component.
 * @param {Boolean} props.required
 *   If true the form control should be a mandatory / required field. This will be
 *   passed from the parent component.
 * @param {Array} props.menuItems
 *   The collection of menu items to be displayed by the form control. This will be
 *   passed from the parent component
 */
const SingleSelectRadioButtonHorizontal = ({
  inputRef,
  uniqueId,
  name,
  variableName,
  defaultValue,
  handleChange,
  updatable,
  required,
  menuItems
}) => {
  // this is required for interact
  useEffect(() => {
    if (inputRef && inputRef.current) {
      inputRef.current.value = defaultValue;
    }
  }, [defaultValue, inputRef]);

  /**
   *
   */
  const handleSingleSelectMenuChange = item => {
    const newValue = {value: item.value}; // TODO: Add more keys to the payload as needed.

    if (newValue.value !== defaultValue) {
      handleChange(newValue);
    }
  };

  return (
    <Styled id="ConfiguratorSingleSelectRadioButtonHorizontal" css={css}>
      <div className="ConfiguratorSingleSelectRadioButtonHorizontal__Container">
        <fieldset className="ConfiguratorSingleSelectMenu__Fieldset">
          <legend className="ConfiguratorSingleSelectMenu__Legend">{name}</legend>
          {menuItems.map((item, index) => {
            return (
              <label htmlFor={`${uniqueId}-${index}`} key={item.value}>
                <Radio
                  id={`${uniqueId}-${index}`}
                  type="radio"
                  value={item.value}
                  onChange={() => handleSingleSelectMenuChange(item)}
                  name={variableName}
                  defaultChecked={defaultValue === item.value}
                  required={required}
                  ref={inputRef}
                  disabled={!updatable}
                />
                <span>{item.displayValue}</span>
              </label>
            );
          })}
        </fieldset>
      </div>
    </Styled>
  );
};

SingleSelectRadioButtonHorizontal.propTypes = {
  /**
   * A reference to the DOM element created in the render method.
   * This will be passed from the parent component.
   */
  inputRef: PropTypes.shape(Object),
  /**
   * A unique id for the rendered DOM element, it should be unique within the HTML
   * document. This will be passed from the parent component.
   */
  uniqueId: PropTypes.string.isRequired,
  /**
   * A display name (label) for the form control. This will be passed from the
   * parent component.
   */
  name: PropTypes.string.isRequired,
  /**
   * The variable name used to connect a form control to its corresponding value in
   * configData . This will be passed from the parent component.
   */
  variableName: PropTypes.string.isRequired,
  /**
   * The initial value for the form control. It is retrieved from the configData
   * object from the configurations repository store. This will be passed from
   * the parent component. See:
   * https://reactjs.org/docs/uncontrolled-components.html#default-values
   */
  defaultValue: PropTypes.string,
  /**
   * The function to be called when a user has changed the value of the form control.
   * This will be passed from the parent component.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * If false the form control should be disabled / read only. This will be
   * passed from the parent component.
   */
  updatable: PropTypes.bool.isRequired,
  /**
   * If true the form control should be a mandatory / required field. This will be
   * passed from the parent component.
   */
  required: PropTypes.bool.isRequired,
  /**
   * The collection of menu items to be displayed by the form control. This will be
   * passed from the parent component
   */
  menuItems: PropTypes.arrayOf(Object).isRequired
};

SingleSelectRadioButtonHorizontal.defaultProps = {
  inputRef: null,
  defaultValue: null
};

export default SingleSelectRadioButtonHorizontal;
