// @oracle-cx-commerce/react-widgets/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {merge} from '@oracle-cx-commerce/utils/generic';

const keys = ['configCssOverrideHelpText', 'configCssOverrideLabel'];

// Default configuration for widgets.
const defaultConfig = {
  properties: [
    {
      id: 'cssOverride',
      type: 'stringType',
      labelResourceId: 'configCssOverrideLabel',
      helpTextResourceId: 'configCssOverrideHelpText'
    }
  ],
  locales: buildConfigResources(resourceBundle, keys)
};

/**
 * It will merge widget specific configuration/settings with default configuration.
 *
 * @param  {Object} config Widget settings
 */
export const mergeDefaultConfig = config => {
  if (!config) {
    return defaultConfig;
  }
  const {properties = [], locales = {}, defaults = {}, ...others} = config;

  return {
    properties: [...defaultConfig.properties, ...properties],
    locales: {...merge(locales, defaultConfig.locales)},
    defaults: {...defaults},
    ...others
  };
};

export default defaultConfig;
