// @oracle-cx-commerce/test/unit/endpoints.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {ERROR} from '@oracle-cx-commerce/test/unit/constants';
import {Response} from 'node-fetch';

/**
 * Calls an endpoint get response and returns the json response
 * @param {*} endpoint the endpoint
 * @param {*} response the endpoint response
 * @param {*} state the application state
 * @param {*} inputPayload the payload sent to the processInput method
 */
export const setupEndpointResponse = async (endpoint, response, state, inputPayload) => {
  response = endpoint.getResponse(response, state, inputPayload);

  return response.getJson();
};

/**
 * Calls an endpoint get request and returns the request
 * @param {*} endpoint the endpoint
 * @param {*} config the endpoint configuration
 */
export const setupEndpointRequest = (endpoint, config, payload = {}, state = {}) => {
  return endpoint.getRequest(config, payload, state);
};

/**
 * Create a dummy error response
 * @param {*} pkg the endpoint
 */
export const createDummyErrorResponse = endpoint => {
  const body = {
    errorCode: '10001',
    message: 'Test message'
  };

  return new Response(JSON.stringify(body), {
    ok: false,
    url: `https://localhost/${endpoint}/`,
    status: ERROR
  });
};
