// @oracle-cx-commerce/test/unit/fetch.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {ERROR} from '@oracle-cx-commerce/test/unit/constants';

// Mock Fetch API
// https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
export class Request {
  constructor(url, options) {
    this.url = url;
    this.options = options;
    this.headers = options.headers ? new Map(Object.entries(options.headers)) : new Map();
  }
}

export class Response {
  constructor({url, status, body, ok, headers}) {
    this.url = url;
    this.status = status;
    this.body = body;
    this.ok = ok;
    this.headers = headers;
  }

  json() {
    return Promise.resolve(this.body);
  }
}

export function fetch(request) {
  return Promise.resolve(new Response({url: request.url, ...request.options}));
}

/**
 * Create a dummy error response
 * @param {*} pkg the endpoint
 */
export const createDummyErrorResponse = endpoint => {
  return new Response({
    ok: false,
    url: `https://localhost/${endpoint}/`,
    body: {
      errorCode: '10001',
      message: 'Test message'
    },
    status: ERROR
  });
};
