// @oracle-cx-commerce/actions/return/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const calculateRefund = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['calculateRefund'],
  description:
    'Calculate the refund for returned items. Optionally takes the x-ccasset-language headers to get translated content in another language. It can also have an optional X-CCOrganization header, which is a JSON field to specify the current organization context of the logged-in user.',
  input: '@oracle-cx-commerce/actions/return/schema/calculate-refund/input.json'
};

export const createReturnRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createReturnRequest'],
  description: 'Creates return request for the order and returns the appropriate return request ID and return details.',
  input: '@oracle-cx-commerce/actions/return/schema/create-return-request/input.json'
};

export const listReturnRequestsForOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listReturnRequestsForOrder'],
  description:
    'Lists all return requests for a given order ID. Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.',
  input: '@oracle-cx-commerce/actions/return/schema/list-return-requests-for-order/input.json'
};

export const getReturnReasons = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getReturnReasons'],
  description: 'This API returns the return reasons configured',
  input: '@oracle-cx-commerce/actions/return/schema/get-return-reasons/input.json'
};

export const getReturnRequest = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getReturnRequest'],
  description:
    'Get return request details for a given returnRequest ID. Optionally takes the x-ccasset-language headers to get translated content in another language. Also it can have optional X-CCOrganization header which is a JSON field to specify current Organization context of logged in user.',
  input: '@oracle-cx-commerce/actions/return/schema/get-return-request/input.json'
};

export const initiateReturn = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['initiateReturn'],
  description: 'This API initiates return for the given order',
  input: '@oracle-cx-commerce/actions/return/schema/initiate-return/input.json'
};

export const validateReturns = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['validateReturns'],
  description:
    'validates Return requests. It validates the return items for each request. If items are not given in requests it gets list of items from an order for which return request can be created.',
  input: '@oracle-cx-commerce/actions/return/schema/validate-returns/input.json'
};
