// @oracle-cx-commerce/actions/profile/index.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createProfile = () => import('@oracle-cx-commerce/actions/profile/actions');
export const createProfileAddress = () => import('@oracle-cx-commerce/actions/profile/actions');
export const deleteProfileAddress = () => import('@oracle-cx-commerce/actions/profile/actions');
export const getCurrentProfile = () => import('@oracle-cx-commerce/actions/profile/actions');
export const getProfileAddress = () => import('@oracle-cx-commerce/actions/profile/actions');
export const getProfileSavedCard = () => import('@oracle-cx-commerce/actions/profile/actions');
export const getShopperSettings = () => import('@oracle-cx-commerce/actions/profile/actions');
export const listProfileAddresses = () => import('@oracle-cx-commerce/actions/profile/actions');
export const listProfileSavedCards = () => import('@oracle-cx-commerce/actions/profile/actions');
export const listProfileSavedCardsForCurrentSite = () => import('@oracle-cx-commerce/actions/profile/actions');
export const listSiteProperties = () => import('@oracle-cx-commerce/actions/profile/actions');
export const removeSavedCardFromProfile = () => import('@oracle-cx-commerce/actions/profile/actions');
export const resetPassword = () => import('@oracle-cx-commerce/actions/profile/actions');
export const saveCardToProfile = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateDefaultProfileAddress = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateExpiredPassword = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateProfile = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateProfileAddress = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateProfileSavedCard = () => import('@oracle-cx-commerce/actions/profile/actions');
export const updateSiteProperties = () => import('@oracle-cx-commerce/actions/profile/actions');
