// @oracle-cx-commerce/actions/profile/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createProfile', 'login'],
  actions: ['login'],
  description:
    'Create Profile. If email does not exist in the repository create a new profile with email and password and return the id else return an error message. Dynamic Properties can also be passed like any other attribute ("dynamicPropertyKey":"dynamicPropertyValue").',
  input: '@oracle-cx-commerce/actions/profile/schema/create-profile/input.json'
};

export const createProfileAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createProfileAddress'],
  description: 'Create profile address. Creates customer profile address.',
  input: '@oracle-cx-commerce/actions/profile/schema/create-profile-address/input.json'
};

export const deleteProfileAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteProfileAddress'],
  description: 'Delete customer profile address.',
  input: '@oracle-cx-commerce/actions/profile/schema/delete-profile-address/input.json'
};

export const getCurrentProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCurrentProfile'],
  description: 'Retrives the profile information of a currently logged in user',
  input: '@oracle-cx-commerce/actions/profile/schema/get-current-profile/input.json'
};

export const getProfileAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProfileAddress'],
  description: 'GET customer profile address.',
  input: '@oracle-cx-commerce/actions/profile/schema/get-profile-address/input.json'
};

export const getProfileSavedCard = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProfileSavedCard'],
  description: 'Get card details for given card id',
  input: '@oracle-cx-commerce/actions/profile/schema/get-profile-saved-card/input.json'
};

export const getShopperSettings = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getShopperSettings'],
  description: 'Retrives global site level shopper settings',
  input: '@oracle-cx-commerce/actions/profile/schema/get-shopper-settings/input.json'
};

export const listProfileSavedCards = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listProfileSavedCards'],
  description: 'List Profile Addresses specific to the current profile.',
  input: '@oracle-cx-commerce/actions/profile/schema/list-profile-saved-cards/input.json'
};

export const listProfileSavedCardsForCurrentSite = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listProfileSavedCardsForCurrentSite'],
  description:
    'This action lists requested cards details for shopper profile. On invocation of this action, all saved cards (profileRepository.savedCards.<profileId>.savedCardsMap) and profile specific saved card ids (profileRepository.savedCards.<profileId>.profileSavedCardIds) will get replaced in application state.',
  input: '@oracle-cx-commerce/actions/profile/schema/list-profile-saved-cards-for-current-site/input.json'
};

export const listSiteProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listSiteProperties'],
  description: 'Gets the list of all the site based properties for the current profile.',
  input: '@oracle-cx-commerce/actions/profile/schema/list-site-properties/input.json'
};

export const removeSavedCardFromProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['removeSavedCardFromProfile'],
  description: 'Remove card for a given card id',
  input: '@oracle-cx-commerce/actions/profile/schema/remove-saved-card-from-profile/input.json'
};

export const resetPassword = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['resetPassword'],
  description: 'Reset Password. Used to generate new password for the user, if profile exists with the entered login.',
  input: '@oracle-cx-commerce/actions/profile/schema/reset-password/input.json'
};

export const saveCardToProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['saveCardToProfile'],
  description: 'Save new card details for the profile',
  input: '@oracle-cx-commerce/actions/profile/schema/save-card-to-profile/input.json'
};

export const updateDefaultProfileAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateDefaultProfileAddress'],
  description: 'Update profile address to set it as the default.',
  input: '@oracle-cx-commerce/actions/profile/schema/update-default-profile-address/input.json'
};

export const updateExpiredPassword = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateExpiredPassword'],
  description: 'Update Expired Password. Used to change password of the profile, if current password is expired.',
  input: '@oracle-cx-commerce/actions/profile/schema/update-expired-password/input.json'
};

export const updateProfile = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateProfile', 'updateSiteProperties'],
  description:
    'Update Profile. Updates customer profile attributes. Dynamic Properties of profile can also be updated like any other attribute ("dynamicPropertyKey":"dynamicPropertyValue").',
  input: '@oracle-cx-commerce/actions/profile/schema/update-profile/input.json'
};

export const updateProfileAddress = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateProfileAddress'],
  description: 'Update profile address. Updates customer profile address attributes.',
  input: '@oracle-cx-commerce/actions/profile/schema/update-profile-address/input.json'
};

export const updateProfileSavedCard = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateProfileSavedCard'],
  description: 'Update card details for given card id',
  input: '@oracle-cx-commerce/actions/profile/schema/update-profile-saved-card/input.json'
};

export const updateSiteProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateSiteProperties'],
  description:
    'Updates the site based properties for the current profile. Can be used to update multiple properties of multiple sites at the same time.',
  input: '@oracle-cx-commerce/actions/profile/schema/update-site-properties/input.json'
};

export const listProfileAddresses = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listProfileAddresses'],
  description: 'List Profile Addresses specific to the current profile.',
  input: '@oracle-cx-commerce/actions/profile/schema/list-profile-addresses/input.json'
};
