// @oracle-cx-commerce/react-widgets/oce/content-listing/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const contentListingConfigResourceKeys = [
  'configContentListingTypeLabel',
  'configContentListingTypeHelpText',
  'configCssOverrideLabel',
  'configCssOverrideHelpText',
  'configItemsPerBlockRequiredHelpText',
  'configItemsPerBlockRequiredLabel',
  'configContentQueryBuilderLabel',
  'configContentListingQueryBuilderText'
];

const contentListingConfigProperties = [
  {
    id: 'contentType',
    type: 'select2Type',
    helpTextResourceId: 'configContentListingTypeHelpText',
    labelResourceId: 'configContentListingTypeLabel',
    dataKey: 'contentTypes',
    required: true
  },
  {
    id: 'blockSize',
    type: 'stringType',
    helpTextResourceId: 'configItemsPerBlockRequiredHelpText',
    labelResourceId: 'configItemsPerBlockRequiredLabel',
    defaultValue: '16',
    required: true
  },
  {
    id: 'queryBuilder',
    type: 'contentQueryBuilderType',
    helpTextResourceId: 'configContentListingQueryBuilderText',
    labelResourceId: 'configContentQueryBuilderLabel',
    required: false
  },
  {
    id: 'cssOverride',
    type: 'stringType',
    labelResourceId: 'configCssOverrideLabel',
    helpTextResourceId: 'configCssOverrideHelpText'
  }
];

const config = mergeDefaultConfig({
  properties: [...contentListingConfigProperties],
  locales: buildConfigResources(resourceBundle, contentListingConfigResourceKeys)
});

export default config;
