// @oracle-cx-commerce/test/config.unit.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

module.exports = {
  verbose: true,
  testURL: 'http://localhost',
  roots: [`${__dirname}`, '<rootDir>/packages', '<rootDir>/scripts/jest'],
  moduleFileExtensions: ['js', 'json', 'jsx', 'node', 'mjs'],
  moduleNameMapper: {'\\.(css|less|scss|sss|styl)$': '@oracle-cx-commerce/test/css-stub.js'},
  testMatch: ['**/__test__/**/*.spec.js?(x)', '**/test/unit/**/*.spec.js?(x)'],
  testPathIgnorePatterns: [
    'packages/tools/cli/templates/(actions|endpoints|fetcher|selectors|utils.js|widget|__test__/expected-files)'
  ],
  setupFiles: [`${__dirname}/unit/setup.js`, '<rootDir>/scripts/jest/setup.js'],
  setupFilesAfterEnv: [`${__dirname}/unit/setup-after-env.js`],
  collectCoverage: true,
  coverageDirectory: '<rootDir>/test-results/unit/coverage',
  coveragePathIgnorePatterns: [
    './packages/plugins/components/react-components-unreleased',
    './packages/plugins/components/react-widgets-unreleased',
    './packages/apps/react-app-extensions',
    // TODO: To be removed once unit test coverage can be improved
    './packages/tools/accelerator-templates',
    './packages/tools/cli/cli-shared',
    './packages/tools/cli/actions',
    './packages/tools/cli/templates',
    './packages/tools/cli/utils',
    './packages/tools/cli/cli.js',
    'tests-temp-dir-.*'
  ],
  coverageThreshold: {
    global: {
      statements: 79,
      branches: 65,
      functions: 77,
      lines: 79
    }
  },
  reporters: [
    'default',
    [
      'jest-junit',
      {
        outputDirectory: 'test-results/unit',
        outputName: 'jest-test-results.xml',
        classNameTemplate: '{filepath}',
        titleTemplate: '{title}',
        ancestorSeparator: ' › ',
        usePathForSuiteName: 'true'
      }
    ]
  ],
  transformIgnorePatterns: ['/node_modules/(?!(@oracle-cx-commerce)/).*/']
};
