// @oracle-cx-commerce/test/config.api.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

module.exports = {
  verbose: true,
  testURL: 'http://localhost',
  roots: ['<rootDir>/packages/tests', '<rootDir>/packages/apps'],
  moduleFileExtensions: ['js', 'jsx', 'json'],
  moduleNameMapper: {'\\.(css|less|scss|sss|styl)$': '@oracle-cx-commerce/test/css-stub.js'},
  transformIgnorePatterns: ['/node_modules/(?!(@oracle-cx-commerce)/).*/'],
  testMatch: ['**/tests/api/**/?(*.)spec.js?(x)'],
  setupFiles: [`${__dirname}/api/jest.api.setup.js`],
  setupFilesAfterEnv: [`${__dirname}/api/jest.api.setup-after-env.js`, 'jest-extended'],
  collectCoverage: true,
  collectCoverageFrom: [
    '<rootDir>/packages/plugins/{actions,endpoints}/**/*.js',
    '<rootDir>/packages/tools/test/api/**/*.js'
  ],
  coveragePathIgnorePatterns: [
    '<rootDir>/packages/plugins/actions/actions/index.js',
    '<rootDir>/packages/plugins/actions/actions/meta.js',
    '<rootDir>/packages/plugins/actions/b2b-actions/index.js',
    '<rootDir>/packages/plugins/actions/b2b-actions/meta.js',
    '<rootDir>/packages/plugins/actions/extensions-actions/index.js',
    '<rootDir>/packages/plugins/actions/extensions-actions/meta.js',
    '<rootDir>/packages/plugins/endpoints/endpoints/index.js',
    '<rootDir>/packages/plugins/endpoints/endpoints/meta.js',
    '<rootDir>/packages/plugins/endpoints/extensions-endpoints/index.js',
    '<rootDir>/packages/plugins/endpoints/extensions-endpoints/meta.js'
  ],
  coverageDirectory: '<rootDir>/test-results/api/coverage',
  /*coverageThreshold: {
    global: {
      statements: 80,
      branches: 64,
      functions: 78,
      lines: 79
    }
  },*/
  reporters: [
    'default',
    [
      'jest-junit',
      {
        outputDirectory: 'test-results/unit',
        outputName: 'jest-test-results.xml',
        classNameTemplate: '{filepath}',
        titleTemplate: '{title}',
        ancestorSeparator: ' › ',
        usePathForSuiteName: 'true'
      }
    ]
  ]
};
