// @oracle-cx-commerce/react-widgets/profile/account-address-book/components/address-book-card/index.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import React from 'react';
import Card from '@oracle-cx-commerce/react-components/card';
import AddressView from '@oracle-cx-commerce/react-widgets/profile/account-address-book/components/address-view';
import MoreAddressActions from '@oracle-cx-commerce/react-widgets/profile/account-address-book/components/more-address-actions';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/profile/account-address-book/components/address-book-card/styles.css';
import PropTypes from 'prop-types';

/**
 * Component to display an account address card.
 * @param {Object} props the properties object
 */
const AddressBookCard = props => {
  const {
    itemId,
    contactInfos = {},
    shippingAddress,
    billingAddress,
    showShippingBadge,
    showBillingBadge,
    showMoreAction,
    itemType
  } = props;

  return (
    <Styled id="AddressBookCard" css={css}>
      {contactInfos && contactInfos[itemId] && (
        <div key={`${itemType}_${itemId}`} className="AddressBookCard__ListItem">
          <Card className="AddressBookCard__Card">
            <div className="AddressBookCard__AddressView">
              <AddressView
                {...props}
                showShippingBadge={showShippingBadge && shippingAddress === itemId}
                showBillingBadge={showBillingBadge && billingAddress === itemId}
                address={contactInfos[itemId]}
                addressInfoClass={'AddressBookCard__Info'}
              />
              {showMoreAction && <MoreAddressActions {...props} />}
            </div>
          </Card>
        </div>
      )}
    </Styled>
  );
};

AddressBookCard.propTypes = {
  /** Address id to display */
  itemId: PropTypes.string.isRequired,

  // eslint-disable-next-line spellcheck/spell-checker
  /**
   * The contactInfos (address) object from redux state(ProfileRepository->contactInfos)
   */
  contactInfos: PropTypes.objectOf(
    PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      address1: PropTypes.string.isRequired,
      city: PropTypes.string.isRequired,
      state: PropTypes.string.isRequired,
      postalCode: PropTypes.string.isRequired,
      country: PropTypes.string.isRequired,
      phoneNumber: PropTypes.string
    })
  ),

  /** Default shipping address id */
  shippingAddress: PropTypes.string,

  /** Default billing address id */
  billingAddress: PropTypes.string,

  /** Flag to show shipping address badge */
  showShippingBadge: PropTypes.bool.isRequired,

  /** Flag to show billing address badge */
  showBillingBadge: PropTypes.bool.isRequired,

  /** Flag to show more address actions */
  showMoreAction: PropTypes.bool.isRequired,

  /** Text to show address type */
  itemType: PropTypes.string.isRequired
};

AddressBookCard.defaultProps = {
  contactInfos: {},
  shippingAddress: '',
  billingAddress: ''
};

export default React.memo(AddressBookCard);
