// @oracle-cx-commerce/react-widgets/profile/profile-account-address-book-summary/components/address-view/index.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import React from 'react';
import Badge from '@oracle-cx-commerce/react-components/badge';
import BuildingIcon from '@oracle-cx-commerce/react-components/icons/building';
import AddressInfo from '@oracle-cx-commerce/react-widgets/profile/profile-address-book/components/address-info';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/profile/profile-account-address-book-summary/components/address-view/styles.css';
import PropTypes from 'prop-types';

/**
 * Component to view address information.
 * @param {Object} props the properties object
 */
const AddressView = props => {
  const {labelDefaultShippingAddress, labelDefaultBillingAddress} = props;
  const {showShippingBadge, showBillingBadge, address, addressInfoClass} = props;

  return (
    address && (
      <Styled id="AddressView" css={css}>
        <React.Fragment>
          <div className="AddressView__Image">
            <BuildingIcon className="AddressView__BuildingIcon" />
          </div>
          <div className={addressInfoClass}>
            <div className="AddressView__Address">
              <address className="AddressView__Address">
                {address.addressType && <div className="AddressView__AddressTitle">{address.addressType}</div>}
                {address.companyName && <div>{address.companyName}</div>}
                <AddressInfo address={address} />
              </address>
            </div>
            {showShippingBadge && <Badge className="AddressView__Badge" badgeText={labelDefaultShippingAddress} />}
            {showBillingBadge && <Badge className="AddressView__Badge" badgeText={labelDefaultBillingAddress} />}
          </div>
        </React.Fragment>
      </Styled>
    )
  );
};

AddressView.propTypes = {
  /** Flag to show default shipping address badge */
  showShippingBadge: PropTypes.bool.isRequired,

  /** Flag to show default shipping address badge */
  showBillingBadge: PropTypes.bool.isRequired,

  /** Default address object */
  address: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    address1: PropTypes.string.isRequired,
    city: PropTypes.string.isRequired,
    state: PropTypes.string.isRequired,
    postalCode: PropTypes.string.isRequired,
    country: PropTypes.string.isRequired,
    phoneNumber: PropTypes.string.isRequired
  }),

  /** Default address object */
  addressInfoClass: PropTypes.string
};

AddressView.defaultProps = {
  address: {},
  addressInfoClass: ''
};

export default AddressView;
