// @oracle-cx-commerce/react-widgets/product/product-add-to-wishlist/components/add-to-wish-list-success-message/index.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import Img from '@oracle-cx-commerce/react-components/img';
import Link from '@oracle-cx-commerce/react-components/link';
import {PAGE_WISH_LIST_DETAILS_LINK} from '@oracle-cx-commerce/commerce-utils/constants';
import PropTypes from 'prop-types';
import React from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/product/product-add-to-wishlist/components/add-to-wish-list-success-message/styles.css';
import {useNumberFormatter} from '@oracle-cx-commerce/react-components/utils/hooks';

export const SuccessMessage = props => {
  const {productSelection, alertPriceUnavailable = ''} = props;
  const {
    productDisplayName = '',
    productSalePrice = 0,
    productThumbImageURL = '',
    productVariantValues = []
  } = productSelection;

  const formatCurrency = useNumberFormatter({style: 'currency'});

  const formatPrice = (price, altPriceMsg) => {
    return price ? formatCurrency(price) : altPriceMsg;
  };

  return (
    <div className="AddToWishListModal__ProductDetailsArea">
      {productThumbImageURL && (
        <Img
          className="AddToWishListModal__ProductImage"
          src={productThumbImageURL}
          alt={productDisplayName}
          title={productDisplayName}
          size="xsmall"
        />
      )}
      <div className="AddToWishListModal__ProductDetails">
        <span className="AddToWishListModal__ProductName">{productDisplayName}</span>
        {productVariantValues && productVariantValues.length > 0 && (
          <div className="AddToWishListModal__ProductVariantDescription">
            {productVariantValues.map(variant => (
              <div key={variant.optionName}>{`${variant.optionName} : ${variant.optionValue}`}</div>
            ))}
          </div>
        )}
      </div>
      <span className="AddToWishListModal__ProductPrice">{formatPrice(productSalePrice, alertPriceUnavailable)}</span>
    </div>
  );
};

/**
 * Component to show success details after a product is added to wishlist
 *
 * @param props
 */
const AddToWishListSuccessMessage = props => {
  const {productSelection = [], wishListId = '', actionViewWishList = ''} = props;

  return (
    <Styled id="AddToWishListSuccessMessage" css={css}>
      <div className="AddToWishListSuccessMessage" role="region" aria-labelledby="AddToWishListModal__Header">
        <SuccessMessage {...props} productSelection={productSelection}></SuccessMessage>
        <Link
          className="AddToWishListSuccessMessage__NewWishList"
          href={`${PAGE_WISH_LIST_DETAILS_LINK}/${wishListId}`}
        >
          {actionViewWishList}
        </Link>
      </div>
    </Styled>
  );
};

AddToWishListSuccessMessage.propTypes = {
  /** List of product information that need to be added to the wishlist */
  productSelection: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.shape({
        /** The id of the product */
        productId: PropTypes.string.isRequired,
        /** The sku id of the variant */
        skuId: PropTypes.string.isRequired,
        /** Display name of the product */
        productDisplayName: PropTypes.string,
        /** Sale price of the product variant */
        productSalePrice: PropTypes.number,
        /** Thumbnail image URI of the product variant */
        productThumbImageURL: PropTypes.string,
        /** Product variant property values */
        productVariantValues: PropTypes.arrayOf(
          PropTypes.shape({
            /** The variant type of the product */
            optionName: PropTypes.string.isRequired,
            /** The variant value of the product */
            optionValue: PropTypes.string.isRequired
          })
        ),
        /** Quantity to be added in the wishlist */
        quantity: PropTypes.number
      }).isRequired
    ).isRequired,
    PropTypes.shape({
      /** The id of the product */
      productId: PropTypes.string.isRequired,
      /** The sku id of the variant */
      skuId: PropTypes.string.isRequired,
      /** Display name of the product */
      productDisplayName: PropTypes.string,
      /** Sale price of the product variant */
      productSalePrice: PropTypes.number,
      /** Thumbnail image URI of the product variant */
      productThumbImageURL: PropTypes.string,
      /** Product variant property values */
      productVariantValues: PropTypes.arrayOf(
        PropTypes.shape({
          /** The variant type of the product */
          optionName: PropTypes.string.isRequired,
          /** The variant value of the product */
          optionValue: PropTypes.string.isRequired
        })
      ),
      /** Quantity to be added in the wishlist */
      quantity: PropTypes.number
    }).isRequired
  ]).isRequired
};

AddToWishListSuccessMessage.defaultProps = {};

export default AddToWishListSuccessMessage;
