// @oracle-cx-commerce/react-widgets/oce/content-item/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

const config = {
  properties: [
    {
      id: 'settings',
      type: 'contentItemSettingsType',
      helpTextResourceId: 'contentItemSettingsText',
      labelResourceId: 'contentItemSettingsLabel',
      required: true
    },
    {
      id: 'cssOverride',
      type: 'stringType',
      labelResourceId: 'cssOverrideLabel',
      helpTextResourceId: 'cssOverrideHelpText'
    }
  ],
  locales: {
    en: {
      resources: {
        contentItemLabel: 'Asset Name',
        contentItemHelpText: 'Name of the asset to be retrieved',
        cssOverrideLabel: 'CSS Class',
        cssOverrideHelpText: 'To override the default widget styling, enter a CSS class name.',
        typeHelpText: 'Source for the content item asset',
        typeLabel: 'Type',
        typeStaticLabel: 'Static',
        typeLayoutLabel: 'Layout',
        typeProductLabel: 'Product Aware',
        productContentTypeLabel: 'Content Type',
        productContentTypeHelpText: 'Select content type to use',
        contentItemSettingsText: '',
        contentItemSettingsLabel: ''
      }
    }
  },

  defaults: {
    media: ''
  }
};
export default config;
