// @oracle-cx-commerce/react-widgets/profile/profile-account-address-book-summary/components/default-account-address-book-view/index.jsx
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import AddressView from '@oracle-cx-commerce/react-widgets/profile/profile-account-address-book-summary/components/address-view';
import Card from '@oracle-cx-commerce/react-components/card';
import Link from '@oracle-cx-commerce/react-components/link';
import {PAGE_ACCOUNT_ADDRESS_BOOK_LINK} from '@oracle-cx-commerce/commerce-utils/constants/page-links';
import PropTypes from 'prop-types';
import React from 'react';
import Styled from '@oracle-cx-commerce/react-components/styled';
import css from '@oracle-cx-commerce/react-widgets/profile/profile-account-address-book-summary/components/default-account-address-book-view/styles.css';

const ACCOUNT_ADDRESS_BOOK = PAGE_ACCOUNT_ADDRESS_BOOK_LINK;

/**
 * Displays default account address (if no default address then first address) View.
 * @param {Object} props the properties object
 */
const DefaultAccountAddressBookView = props => {
  const {actionManage, labelManage} = props;

  const {defaultShippingAddress = {}, defaultBillingAddress = {}} = props;

  return (
    <Styled id="DefaultAccountAddressBookView" css={css}>
      <div className="DefaultAccountAddressBookView__Card">
        <Card>
          <div className="DefaultAccountAddressBookView__DefaultAddresses">
            {defaultShippingAddress.repositoryId && (
              <div className="DefaultAccountAddressBookView__AddressView">
                <AddressView
                  {...props}
                  showShippingBadge={true}
                  showBillingBadge={defaultShippingAddress.repositoryId === defaultBillingAddress.repositoryId}
                  address={defaultShippingAddress}
                />
              </div>
            )}
            {defaultBillingAddress.repositoryId &&
              defaultShippingAddress.repositoryId !== defaultBillingAddress.repositoryId && (
                <div className="DefaultAccountAddressBookView__AddressView">
                  <AddressView
                    {...props}
                    showShippingBadge={false}
                    showBillingBadge={true}
                    address={defaultBillingAddress}
                  />
                </div>
              )}
          </div>
          <div className="DefaultAccountAddressBookView__RightContent">
            <Link href={ACCOUNT_ADDRESS_BOOK} aria-label={labelManage}>
              {actionManage}
            </Link>
          </div>
        </Card>
      </div>
    </Styled>
  );
};

DefaultAccountAddressBookView.propTypes = {
  /** Default shipping address object pulled from redux state */
  defaultShippingAddress: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    address1: PropTypes.string,
    city: PropTypes.string,
    state: PropTypes.string,
    postalCode: PropTypes.string,
    country: PropTypes.string,
    phoneNumber: PropTypes.string
  }),

  /** Default billing address object pulled from redux state */
  defaultBillingAddress: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    address1: PropTypes.string,
    city: PropTypes.string,
    state: PropTypes.string,
    postalCode: PropTypes.string,
    country: PropTypes.string,
    phoneNumber: PropTypes.string
  })
};

DefaultAccountAddressBookView.defaultProps = {
  defaultShippingAddress: {},
  defaultBillingAddress: {}
};

export default DefaultAccountAddressBookView;
