// @oracle-cx-commerce/actions/share-setting/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const createShareSetting = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createShareSetting'],
  description: 'This operation is used to create a share setting for a resource.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/create-share-setting/input.json'
};

export const listShareSettings = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listShareSettings'],
  description: 'This operation is used to get the collection of share settings from Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/list-share-settings/input.json'
};

export const getShareSetting = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getShareSetting'],
  description: 'This operation is used to get the share setting information from Oracle Commerce Cloud.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/get-share-setting/input.json'
};

export const updateShareSetting = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateShareSetting'],
  description: 'This operation is used to update a share setting for a resource.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/update-share-setting/input.json'
};

export const deleteShareSetting = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteShareSetting'],
  description: 'This operation is to delete the share setting information of purchase list.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/delete-share-setting/input.json'
};

export const createEmailConfigs = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createEmailConfigs'],
  description: 'Create Email Configs for a given Share Setting.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/create-email-configs/input.json'
};

export const updateEmailConfigs = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateEmailConfigs'],
  description: 'Update Email Configs for a given Share Setting.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/update-email-configs/input.json'
};

export const deleteEmailConfigs = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteEmailConfigs'],
  description: 'Delete Email Configs for a given Share Setting.',
  input: '@oracle-cx-commerce/actions/share-setting/schema/delete-email-configs/input.json'
};
