// @oracle-cx-commerce/react-widgets/profile/add-or-edit-account-contact/meta.js
/*
 ** Copyright (c) 2019 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'admin',
  'accountAddressManager',
  'approver',
  'profileAddressManager',
  'buyer',
  'actionSave',
  'labelCancel',
  'labelRemove',
  'labelFirstName',
  'labelLastName',
  'labelEmailAddress',
  'headingEditContact',
  'headingAddContact',
  'textActive',
  'textContactRemovalWarning',
  'textConfirmContactRemoval',
  'actionConfirm',
  'labelRemoveContact',
  'headingConfirmContactRemoval',
  'textAllFieldsRequired',
  'labelMultipleAccounts',
  'labelYes',
  'labelNo',
  'labelRoles',
  'labelAriaCheckboxCheckedReadOnly',
  'messageContactCreationSuccess',
  'messageContactSaved',
  'messageContactRemoved',
  'closeLinkAltText'
];
export const AddOrEditAccountContact = {
  packageId: '@oracle-cx-commerce/react-widgets',
  config: defaultConfig,
  resources: buildResources(resourceBundle, widgetResourceKeys),
  availableToAllPages: false,
  pageTypes: ['profile']
};
