// @oracle-cx-commerce/react-widgets/category/collection-navigation/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const configResourceKeys = [
  'configViewAllOptionHelpText',
  'configViewAllOptionLabel',
  'configNumberOfChildCategoriesToDisplayHelpText',
  'configNumberOfChildCategoriesToDisplayLabel'
];

const config = mergeDefaultConfig({
  properties: [
    {
      id: 'viewAllOption',
      type: 'booleanType',
      name: 'viewAllOption',
      helpTextResourceId: 'configViewAllOptionHelpText',
      labelResourceId: 'configViewAllOptionLabel',
      defaultValue: true,
      required: false
    },
    {
      id: 'numberOfChildCategoriesToDisplay',
      type: 'stringType',
      name: 'numberOfChildCategoriesToDisplay',
      helpTextResourceId: 'configNumberOfChildCategoriesToDisplayHelpText',
      labelResourceId: 'configNumberOfChildCategoriesToDisplayLabel',
      minLength: 1,
      maxLength: 1,
      defaultValue: '3',
      pattern: '[1-9]',
      required: true
    }
  ],
  locales: buildConfigResources(resourceBundle, configResourceKeys)
});

export default config;
