// @oracle-cx-commerce/test/api/logger.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

const winston = require('winston');
const path = require('path');
const {enumerateErrorFormat, consoleFormatter} = require('@oracle-cx-commerce/logger/common');

const {createLogger, format} = winston;
const {combine, timestamp, label, json, splat} = format;

const fileFormat = combine(
  label({label: `${process.pid}`}),
  timestamp({format: 'YYYY-MM-DD HH:mm:ss'}),
  splat(),
  json()
);

const loggingOption = {
  console: {
    level: 'info',
    handleExceptions: true,
    json: false,
    colorize: true
  },
  file: {
    level: 'debug',
    filename: path.join(process.cwd(), 'test-results', 'api', `api-test-${process.pid}.log`),
    handleExceptions: true,
    format: fileFormat,
    maxsize: 5242880, // 5MB
    maxFiles: 5
  }
};

// Creates a logger singleton specifically for cli tools running locally
// Does not save those log messages into a file
const log = createLogger({
  level: 'debug',
  format: combine(enumerateErrorFormat(), format.timestamp({format: 'YYYY-MM-DD HH:mm:ss'})),
  transports: [
    new winston.transports.Console({
      ...loggingOption.console,
      format: consoleFormatter(`api-test-${process.pid}`)
    }),
    new winston.transports.File(loggingOption.file)
  ]
});

export default log;
