// @oracle-cx-commerce/actions/scheduled-order/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const convertCartToScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['convertCartToScheduledOrder'],
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/convert-cart-to-scheduled-order/input.json'
};

export const createScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['createScheduledOrder'],
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/create-scheduled-order/input.json',
  description: 'Create new scheduled order with the given schedule information.'
};

export const deleteScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['deleteScheduledOrder'],
  description: 'Deletes the scheduled order with the specified scheduled order id.',
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/delete-scheduled-order/input.json'
};

export const getScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getScheduledOrder'],
  description: 'Get scheduled order retrieves the scheduled order for the given id.',
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/get-scheduled-order/input.json'
};

export const listExecutionStatusForScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listExecutionStatusForScheduledOrder'],
  description: 'Retrieves the execution status for the given scheduled order id.',
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/list-execution-status-for-scheduled-order/input.json'
};

export const listScheduledOrders = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['listScheduledOrders'],
  description: 'Retrieves the scheduled orders associated with the logged in user.',
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/list-scheduled-orders/input.json'
};

export const updateScheduledOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateScheduledOrder'],
  description: 'Updates the scheduled order configuration.',
  input: '@oracle-cx-commerce/actions/scheduled-order/schema/update-scheduled-order/input.json'
};
