// @oracle-cx-commerce/react-widgets/profile/account-contacts/meta.js
/*
 ** Copyright (c) 2019 Oracle and/or its affiliates.
 */
import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import config from '@oracle-cx-commerce/react-widgets/profile/account-contacts/config';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'admin',
  'accountAddressManager',
  'approver',
  'profileAddressManager',
  'buyer',
  'emailasc',
  'emaildesc',
  'firstNameasc',
  'firstNamedesc',
  'lastNameasc',
  'lastNamedesc',
  'headingAccountContacts',
  'labelNewContact',
  'labelAriaSelectRole',
  'labelAriaSortBy',
  'labelAriaListOf',
  'labelAriaTextFilter',
  'labelAriaSelectItemsPerPage',
  'textItemsPerPage',
  'textFilterByNameOrEmail',
  'textActive',
  'textSearchResult',
  'labelLoadMore',
  'textInactive',
  'textAllRoles',
  'textRetrievingAccountContacts',
  'closeLinkAltText',
  'textNoAccountContactsAvailable',
  'labelLoadingAccountContacts',
  'searchIconAltText'
];

export const AccountContacts = {
  packageId: '@oracle-cx-commerce/react-widgets',
  config: mergeDefaultConfig(config),
  availableToAllPages: false,
  pageTypes: ['profile'],
  resources: buildResources(resourceBundle, widgetResourceKeys),
  actions: ['listOrganizationMembers', 'getOrganization']
};
