// @oracle-cx-commerce/react-widgets/profile/account-contacts/utils.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {sortArrayByKeyAsc, t} from '@oracle-cx-commerce/utils/generic';

export const ADMIN = 'admin';
export const allRoles = ['admin', 'accountAddressManager', 'approver', 'profileAddressManager', 'buyer'];
export const defaultRole = {function: 'All Roles', repositoryId: 'All'};
const sortOptions = [
  {
    sortTerm: 'email',
    order: 'asc'
  },
  {
    sortTerm: 'email',
    order: 'desc'
  },
  {
    sortTerm: 'firstName',
    order: 'asc'
  },
  {
    sortTerm: 'firstName',
    order: 'desc'
  },
  {
    sortTerm: 'lastName',
    order: 'asc'
  },
  {
    sortTerm: 'lastName',
    order: 'desc'
  }
];

/**
 * The method returns a sorted array containing localized values for sort options.
 * @param {*} locales - object containing translations for the sort options
 */
export const getSortOptions = locales => {
  const localisedSortOptions = sortOptions.map(sortChoice => {
    const option = {...sortChoice};
    option['display'] = locales[`${sortChoice.sortTerm}${sortChoice.order}`];

    return option;
  });
  sortArrayByKeyAsc(localisedSortOptions, 'display');

  return localisedSortOptions;
};

/**
 * The method returns a sorted array containing localized values for roles.
 * @param {*} roles - array containing role objects
 * @param {*} localeTranslations - object containing translations for roles
 */
export const convertRolesToDisplayableOptions = (roles, localeTranslations) => {
  //map role functions to displayable options

  const localisedRoles = roles.map(roleObj => {
    return {...roleObj, display: localeTranslations[roleObj.function]};
  });
  sortArrayByKeyAsc(localisedRoles, 'display');

  return localisedRoles;
};

/*
 **The method converts the itemsPerPage comma separated string into object with localized values for display
 */
export const populateItemsPerPage = (itemsPerPageString, text) => {
  return itemsPerPageString.split(',').map(item => {
    return {
      value: item.trim(),
      display: `${t(text, {number: item.trim()})}`
    };
  });
};
