// @oracle-cx-commerce/react-widgets/checkout/checkout-invoice/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'actionCancel',
  'buttonSaveAndContinue',
  'closeLinkAltText',
  'headingAddBillingAddress',
  'headingBillingAddress',
  'headingEditBillingAddress',
  'labelAddANewAddress',
  'labelAddressBook',
  'labelPurchaseOrderNumber',
  'labelCancel',
  'labelCompanyName',
  'labelCountry',
  'labelDefaultAddress',
  'labelDefaultBillingAddress',
  'labelDefaultShippingAddress',
  'labelEditAddress',
  'labelFirstName',
  'labelLastName',
  'labelNickName',
  'labelNoAccountAddressesAvailable',
  'labelNoDefaultAddressesAvailable',
  'labelNoInheritedAddressesAvailable',
  'labelNoProfileAddressesAvailable',
  'labelOpenAddressBook',
  'labelInvoice',
  'labelPhoneNumberOptional',
  'labelSaveAsANewAccountAddress',
  'labelSaveAsANewProfileAddress',
  'textAccountAddressBook',
  'textAddressesForThisAccount',
  'textAllFieldsRequired',
  'textDefaultAddresses',
  'textEnterABillingAddress',
  'textInheritedAddresses',
  'textLoading',
  'textLoadMoreAccountAddress',
  'textLoadMoreInheritedAddress',
  'textLoadMoreProfileAddress',
  'textOpenAddressBookAndChooseBillingAddress',
  'textProfileAddresses',
  'textRequiredField',
  'textInvalidField',
  'textUseThisAddress',
  'labelState',
  'labelStreetAddress',
  'labelTownCity',
  'labelZipCode'
];

export const CheckoutInvoice = {
  packageId: '@oracle-cx-commerce/react-widgets',
  availableToAllPages: false,
  pageTypes: ['checkout-payment', 'pending-payment'],
  requiresContext: ['payment_context'],
  config: defaultConfig,
  resources: buildResources(resourceBundle, widgetResourceKeys)
};
