// @oracle-cx-commerce/react-widgets/checkout/checkout-invoice/selectors.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {getCurrentOrder, getCurrentProfile, getOrder, getPage} from '@oracle-cx-commerce/commerce-utils/selector';
import {
  getPaymentGroupsByTypes,
  isPaymentEnabledForApproval,
  isPaymentTypeEnabled,
  isPaymentDisabled,
  getPaymentEnabledForScheduledOrder
} from '@oracle-cx-commerce/react-components/utils/payment';
import {PAYMENT_TYPE_INVOICE, PAYMENT_METHOD_INVOICE_REQUEST} from '@oracle-cx-commerce/commerce-utils/constants';

/**
 * Returns widget's required data
 * @param {Object} state the state object
 * @returns required data
 */
export const getComponentData = state => {
  // selectors
  const contextOrderId = getPage(state).contextId;
  const order = contextOrderId ? getOrder(state, {id: contextOrderId}) : getCurrentOrder(state);

  return {
    isPaymentDisabled: isPaymentDisabled(order),
    isDisplayCheckoutInvoice: isPaymentTypeEnabled(state, PAYMENT_TYPE_INVOICE),
    appliedInvoicePaymentGroups: getPaymentGroupsByTypes(order, [PAYMENT_METHOD_INVOICE_REQUEST]),
    isPaymentMethodEnabledForApproval: isPaymentEnabledForApproval(state, PAYMENT_TYPE_INVOICE),
    isApprovalEnabled: getCurrentProfile(state).derivedApprovalRequired,
    PaymentInfoForScheduledOrder: getPaymentEnabledForScheduledOrder(state, PAYMENT_TYPE_INVOICE)
  };
};
