// @oracle-cx-commerce/react-widgets/product/product-delivery-options/config.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildConfigResources} from '@oracle-cx-commerce/resources/utils';
import {mergeDefaultConfig} from '@oracle-cx-commerce/react-widgets/config';

export const productDeliveryOptionsConfigResourceKeys = [
  'configNumberOfMatchingStoresToDisplayHelpText',
  'configNumberOfMatchingStoresToDisplayLabel',
  'configDisplayNumberOfItemsAvailableHelpText',
  'configDisplayNumberOfItemsAvailableLabel',
  'configDisplayAvailabilityDateHelpText',
  'configDisplayAvailabilityDateLabel'
];

export const productDeliveryOptionsConfigProperties = [
  {
    id: 'numberOfMatchingStoresToDisplay',
    type: 'stringType',
    name: 'numberOfMatchingStoresToDisplay',
    helpTextResourceId: 'configNumberOfMatchingStoresToDisplayHelpText',
    labelResourceId: 'configNumberOfMatchingStoresToDisplayLabel',
    defaultValue: '10',
    required: true,
    maxLength: 3
  },
  {
    id: 'displayNumberOfItemsAvailable',
    type: 'booleanType',
    name: 'displayNumberOfItemsAvailable',
    helpTextResourceId: 'configDisplayNumberOfItemsAvailableHelpText',
    labelResourceId: 'configDisplayNumberOfItemsAvailableLabel',
    defaultValue: false
  },
  {
    id: 'displayAvailabilityDate',
    type: 'booleanType',
    name: 'displayAvailabilityDate',
    helpTextResourceId: 'configDisplayAvailabilityDateHelpText',
    labelResourceId: 'configDisplayAvailabilityDateLabel',
    defaultValue: false
  }
];

const productDeliveryOptionsConfig = mergeDefaultConfig({
  properties: productDeliveryOptionsConfigProperties,
  locales: buildConfigResources(resourceBundle, productDeliveryOptionsConfigResourceKeys)
});

export default productDeliveryOptionsConfig;
