// @oracle-cx-commerce/react-widgets/product/product-delivery-options/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import productDeliveryOptionsConfig from '@oracle-cx-commerce/react-widgets/product/product-delivery-options/config';

export const productDeliveryOptionsWidgetResourceKeys = [
  'actionChangeStore',
  'actionSelectStore',
  'closeLinkAltText',
  'headingStoreLocator',
  'labelFindAStore',
  'labelPickUpInStore',
  'labelHomeDelivery',
  'textStoreSearchHelper',
  'textStoresFilter',
  'textSelect',
  'textInStock',
  'textPreOrderable',
  'textBackOrderable',
  'textOutOfStock',
  'textInStockWithQty',
  'textBackOrderableWithQty',
  'textPreOrderableWithQty',
  'textBackOrderableWithAvailabilityDate',
  'textPreOrderableWithAvailabilityDate',
  'textPreOrderableWithQtyAndDate',
  'textBackOrderableWithQtyAndDate',
  'textStoreSearchResults',
  'messageInsufficientStockAtStore'
];
export const ProductDeliveryOptions = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, productDeliveryOptionsWidgetResourceKeys),
  config: productDeliveryOptionsConfig,
  availableToAllPages: false,
  pageTypes: ['product'],
  requiresContext: ['product_context', 'container_context']
};
