// @oracle-cx-commerce/react-widgets/profile/add-to-purchase-list/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'labelAddToPurchaseList',
  'labelAddSelectedToPurchaseList',
  'closeLinkAltText',
  'headerLoginOrCreateAccount',
  'labelAddToPurchaseListNotLoggedIn',
  'actionLogin',
  'actionCreateAnAccount',
  'headerAddToPurchaseList',
  'actionAdd',
  'actionCreateNewPurchaseList',
  'headerCreatePurchaseList',
  'alertDuplicatePurchaseList',
  'actionSave',
  'actionCancel',
  'labelListName',
  'headerAddedToPurchaseList',
  'actionViewPurchaseList',
  'alertPriceUnavailable',
  'messageAddedToPurchaseList',
  'labelAddToPurchaseListName'
];

export const AddToPurchaseList = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: defaultConfig,
  requiresContext: ['product_selection_context']
};
