// @oracle-cx-commerce/react-widgets/product/product-add-to-wishlist/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';

import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

export const widgetResourceKeys = [
  'labelAddToWishList',
  'headerAddToWishList',
  'headerAddedToWishList',
  'actionViewWishList',
  'headerCreateWishList',
  'actionCreateNewWishList',
  'headerLoginOrCreateAccount',
  'labelListName',
  'closeLinkAltText',
  'labelAddSelectedToPurchaseList',
  'actionLogin',
  'actionCreateAnAccount',
  'alertPriceUnavailable',
  'alertDuplicateWishList',
  'alertItemAlreadyInWishlist',
  'wishlistAccessLevelPrivateHelpText',
  'wishlistAccessLevelGroupHelpText',
  'wishlistAccessLevelSharedHelpText',
  'alertItemAlreadyAddedToWishlist',
  'headingWishlistPrivacySettings',
  'textPrivate',
  'textGroup',
  'textShared',
  'labelAddToWishListNotLoggedIn',
  'alertMoveToWishListSuccess',
  'actionAdd',
  'actionSave',
  'actionCancel',
  'messageAddedToPurchaseList',
  'labelAddToWishListName',
  'textCreateWishListDescription'
];

export const ProductAddToWishList = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: defaultConfig,
  requiresContext: ['product_selection_context']
};
