// @oracle-cx-commerce/react-widgets/product/product-add-to-wishlist/utils.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import {isEmptyObject} from '@oracle-cx-commerce/utils/generic';
import {getProduct, getSku} from '@oracle-cx-commerce/commerce-utils/selector';

/**
 * Computes the payload for `addProductToWishlist` from product selection data
 * @param {Object} productSelection The product selection passed to the component
 * @returns {Object} Computed payload
 */
export function getUpdateWishListItems(productSelection) {
  let items = [];
  if (typeof productSelection === 'object') {
    items = {
      productTitle: productSelection.productDisplayName,
      productPrice: productSelection.productSalePrice,
      productSalePrice: productSelection.productSalePrice,
      productUrl: productSelection.productThumbImageURL,
      productProductId: productSelection.productId,
      productSkuId: productSelection.skuId,
      productVariantOptions: productSelection.productVariantValues,
      productCurrencyCode: productSelection.USD,
      notificationProductPrice: productSelection.notificationProductPrice,
      productImageUrl: productSelection.productThumbImageURL,
      productDesiredQuantity: productSelection.quantity ? productSelection.quantity : 1
    };
  }

  return items;
}

export const getDetailsFromProductAndSku = (product, sku, quantity) => {
  const skuId = sku.repositoryId,
    {variantToSkuLookup, variantOptionsSeed} = product,
    variantDetails = [];
  let thumbImageURLs = [null];

  // Thumbnail image
  if (sku.thumbImageURLs && sku.thumbImageURLs.length) {
    ({thumbImageURLs} = sku);
  } else if (product.thumbImageURLs && product.thumbImageURLs.length) {
    ({thumbImageURLs} = product);
  }

  // Variant details
  const optionsStr = Object.entries(variantToSkuLookup).find(item => item[1] === skuId)[0];
  (optionsStr ? optionsStr : '').split(';').forEach(option => {
    if (option) {
      const [key, value] = option.split('=');

      const optionDetails = variantOptionsSeed[key],
        {options} = optionDetails;

      const {optionName} = optionDetails;
      const optionValue = options.find(item => item.value === parseInt(value, 10))['name'];

      variantDetails.push({optionName, optionValue});
    }
  });

  return {
    productId: product.id,
    skuId: sku.repositoryId,
    productDisplayName: sku.displayName ? sku.displayName : product.displayName,
    productSalePrice: sku.salePrice ? sku.salePrice : sku.listPrice,
    productThumbImageURL: thumbImageURLs[0],
    productVariantValues: variantDetails,
    quantity
  };
};

function getDetailsFromCommerceItem(commerceItem) {
  return {
    productId: commerceItem.productId,
    skuId: commerceItem.catRefId,
    productDisplayName: commerceItem.displayName,
    productSalePrice: commerceItem.salePrice ? commerceItem.salePrice : commerceItem.listPrice,
    productThumbImageURL: commerceItem.primaryThumbImageURL,
    productVariantValues: commerceItem.variant,
    quantity: commerceItem.quantity
  };
}

/**
 * Computes details of products from the product selection data
 * @param {Object} state The Redux state
 * @param {Object} item A product selection item, consisting of either combination of product and sku ids, or a commerce item
 * @returns {Object} the product details comprising `productId`, `skuId`, `productDisplayName`, `productThumbImageURL`, etc.
 */
export function getProductDetails(state, item) {
  const {productId, skuId, quantity, commerceItem} = item;

  if (productId && skuId) {
    const product = getProduct(state, {productId});
    const sku = getSku(state, {skuId});

    if (!isEmptyObject(product) && !isEmptyObject(sku)) {
      return getDetailsFromProductAndSku(product, sku, quantity);
    }
  } else if (commerceItem) {
    if (!isEmptyObject(commerceItem)) {
      return getDetailsFromCommerceItem(commerceItem);
    }
  }

  return {};
}
