// @oracle-cx-commerce/react-widgets/profile/profile-purchase-list-action-buttons/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';
import {buildResources} from '@oracle-cx-commerce/resources/utils';

const widgetResourceKeys = [
  'closeLinkAltText',
  'textShareList',
  'textEditList',
  'textCopyList',
  'textDeleteList',
  'textEditPurchaseList',
  'textAllFieldsRequired',
  'labelListName',
  'labelDescriptionOptional',
  'actionSave',
  'actionCancel',
  'actionConfirm',
  'textDeletePurchaseList',
  'textConfirmDeletePurchaseListMessage',
  'actionShowMore',
  'textActionButtonsSeparator',
  'alertPurchaseListUpdated',
  'alertPurchaseListDeleted',
  'alertPurchaseListCreated',
  'textCopyPurchaseList',
  'labelCopy',
  'textSharePurchaseList',
  'labelEmailAddress',
  'labelAddComment',
  'labelShareWithAccount',
  'labelShareWithEmailAddress',
  'textAddEmailAddress',
  'labelPermissions',
  'textCanEdit',
  'textReadOnly',
  'textRemoveShopper',
  'alertPurchaseListSharingUpdated'
];

export const ProfilePurchaseListActionButtons = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  config: defaultConfig,
  availableToAllPages: false,
  pageTypes: ['profile'],
  requiresContext: ['container_context']
};
