// @oracle-cx-commerce/actions/quote/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const requestQuote = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['requestQuote'],
  description: 'Request Quote. Request the quote on incomplete order',
  input: '@oracle-cx-commerce/actions/quote/schema/request-quote/input.json'
};

export const rejectQuote = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['rejectQuote'],
  description: 'Reject Quote. Reject the quote on the incomplete order',
  input: '@oracle-cx-commerce/actions/quote/schema/reject-quote/input.json'
};

export const submitQuotedOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['submitQuotedOrder'],
  description: 'Submits the quoted order from the given id in the path.',
  input: '@oracle-cx-commerce/actions/quote/schema/submit-quoted-order/input.json'
};

export const updateQuotedOrder = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['updateQuotedOrder'],
  description: 'Updates the quoted order from the given id in the path.',
  input: '@oracle-cx-commerce/actions/quote/schema/update-quoted-order/input.json'
};
