// @oracle-cx-commerce/react-widgets/profile/profile-scheduled-order-summary-information/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

import * as resourceBundle from '@oracle-cx-commerce/resources';
import {buildResources} from '@oracle-cx-commerce/resources/utils';
import defaultConfig from '@oracle-cx-commerce/react-widgets/config';

const widgetResourceKeys = [
  'textScheduleId',
  'textScheduleFrequency',
  'textScheduleLastOrderDate',
  'textScheduleNextOrderDate',
  'textScheduleStatus',
  'textApproverName',
  'textApproverComments',
  'textApproverReasons',
  'textPeriodicSchedule',
  'textCalendarSchedule',
  'textDailySchedule',
  'textWeeklySchedule',
  'textMonthlySchedule',
  'textCustomSchedule',
  'textOrderStatus'
];

export const ProfileScheduledOrderSummaryInformation = {
  packageId: '@oracle-cx-commerce/react-widgets',
  resources: buildResources(resourceBundle, widgetResourceKeys),
  availableToAllPages: false,
  pageTypes: ['Scheduled Order Details'],
  requiresContext: ['container_context', 'order_context'],
  config: defaultConfig
};
