// @oracle-cx-commerce/actions/custom-properties/meta.js
/*
 ** Copyright (c) 2020 Oracle and/or its affiliates.
 */

export const getCartItemCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCartItemCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-cart-item-custom-properties/input.json',
  description: 'Get CommerceItem custom properties.'
};

export const getContactInfoCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getContactInfoCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-contactInfo-custom-properties/input.json',
  description: 'Get ContactInfo custom properties.'
};

export const getOrderCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getOrderCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-order-custom-properties/input.json',
  description: 'Get Order custom properties.'
};

export const getProductCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getProductCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-product-custom-properties/input.json',
  description: 'Get Product custom properties.'
};

export const getUserCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getUserCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-user-custom-properties/input.json',
  description: 'Get User custom properties.'
};

export const getCategoryCustomProperties = {
  packageId: '@oracle-cx-commerce/actions',
  endpoints: ['getCategoryCustomProperties'],
  input: '@oracle-cx-commerce/actions/custom-properties/schema/get-category-custom-properties/input.json',
  description: 'Get Collection custom properties.'
};
